---
title: "CD33"

knit: (function(input, ...) {
    rmarkdown::render(
      input,
      output_format = "html_document",
      output_dir = "../results/html_docs",
    )
  })
---

<style>
  .scrollingplot{
      overflow-x:scroll;
      white-space: nowrap;
  }

  .scrollingplot img{
     max-width: 40cm;
  }

</style>

```{r, eval = TRUE, echo = FALSE, warning = FALSE, message = FALSE}
setwd("../../")
source("cartcontent/scr/01_functions.R")
```

<h4>Description</h4>
Myeloid cell surface antigen CD33 ([CD33](https://www.uniprot.org/uniprot/P20138)). Sialic-acid-binding immunoglobulin-like lectin (Siglec) that plays a role in mediating cell-cell interactions and in maintaining immune cells in a resting state. 

<br> 

<h4>Clinical trials</h4> 
[NCT03638206](https://clinicaltrials.gov/ct2/show/NCT03638206) - Esophagus Cancer, Renal Carcinoma, Hepatoma, Lung Cancer, Mesothelioma, Ovarian Cancer, Pancreatic Cancer, Melanoma

<br>

<h4>Target gene expression</h4>
The following plot describes the expression, in different tissues and TCGA cancers, of the different isoforms of the gene CD33.
<div class="scrollingplot">
```{r , echo = FALSE, warning = FALSE, message = FALSE, fig.width = 80, fig.height = 40}
setwd("../../")
boxplot_isoforms_all_tissues(hgnc = "CD33") # it creates the boxplot in the html and saves it as .pdf in the plot/isoform_expression folder
```
</div>

<br>

<h4>Target essentiality</h4>
```{r, eval = TRUE, echo = FALSE, warning = FALSE, message = F, fig.width = 25, fig.height= 15}
setwd("../../")
target_essentiality_fun(gene_name_hgnc = "CD33") # it creates the boxplot in the html and saves it as .pdf in the plot/target_essentiality folder
```

<br>

<h4>Protein properties</h4>
The following plot describes the target topology, highlighting the position of the endo-, ecto- and transmembrane domains, as well as the position of a signal peptide, when present. 
```{r, eval = TRUE, echo = FALSE, warning = FALSE, message = F, fig.width = 10, fig.height = 2}
setwd("../../")
plot_function("CD33") # it creates the plot in the html and saves it as .pdf in the plot/protein_properties folder
```