---
title: "CLDN18.2"

knit: (function(input, ...) {
    rmarkdown::render(
      input,
      output_format = "html_document",
      output_dir = "../results/html_docs",
    )
  })
---
<style>
  .scrollingplot{
      overflow-x:scroll;
      white-space: nowrap;
  }

  .scrollingplot img{
     max-width: 40cm;
  }

</style>

```{r, eval = TRUE, echo = FALSE, warning = FALSE, message = FALSE}
setwd("../../")
source("cartcontent/scr/00_functions.R")
```

<h4>Description</h4>
Claudin-18, isoform 2 ([CLDN18.2](https://www.uniprot.org/uniprot/P56856)). Plays a major role in tight junction-specific obliteration of the intercellular space, through calcium-independent cell-adhesion activity.

<br> 

<h4>Clinical trials</h4> 
[NCT04404595](https://clinicaltrials.gov/ct2/show/NCT04404595) - Pancreatic Cancer <br>
[NCT03159819](https://clinicaltrials.gov/ct2/show/NCT03159819) - Pancreatic Cancer

<br>

<h4>Target gene expression</h4>
The following plot describes the expression, in different tissues and TCGA cancers, of the different isoforms of the gene CLDN18.
<div class="scrollingplot">
```{r , echo = FALSE, warning = FALSE, message = FALSE, fig.width = 80, fig.height = 35}
setwd("../../")
boxplot_isoforms_all_tissues(hgnc = "CLDN18") # it creates the boxplot in the html and saves it as .pdf in the plot/isoform_expression folder
```
</div>
<br>

<h4>Target essentiality</h4>
```{r, eval = TRUE, echo = FALSE, warning = FALSE, message = F, fig.width = 25, fig.height= 12}
setwd("../../")
target_essentiality_fun(gene_name_hgnc = "CLDN18") # it creates the boxplot in the html and saves it as .pdf in the plot/target_essentiality folder
```

<br>

<h4>Protein properties</h4>
The following plot describes the target topology, highlighting the position of the endo-, ecto- and transmembrane domains, as well as the position of a signal peptide, when present. 
The last three rows describe the position of potential epitopes, as well as phosphorylation and glycosylation sites. 
```{r, eval = TRUE, echo = FALSE, warning = FALSE, message = F, fig.width = 10, fig.height= 4}
setwd("../../")
plot_function("CLDN18_2") # it creates the plot in the html and saves it as .pdf in the plot/protein_properties folder
```
