<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Contributors: Lars Rønn Olsen, Christina Bligaard Pedersen, Mike Leipold, and Holden Maecker" />


<title>Data scientist’s primer to analysis of mass cytometry data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">The Cancer Omics group</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="projects.html">Projects</a>
    </li>
    <li>
      <a href="publications.html">Publications</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Teaching
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://kurser.dtu.dk/course/22140">Introduction to Systems Biology</a>
    </li>
    <li>
      <a href="spectracular.html">Bioinformatics for Precision Medicine</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    People
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="group_members.html">Group members</a>
    </li>
    <li>
      <a href="collaborators.html">Collaborators</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Resources
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="cytof.html">CyTOF primers</a>
    </li>
    <li>
      <a href="spectracular.html">Spectracular</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Data scientist’s primer to analysis of mass cytometry data</h1>
<h4 class="author">Contributors: Lars Rønn Olsen, Christina Bligaard Pedersen, Mike Leipold, and Holden Maecker</h4>

</div>


<div id="preamble" class="section level1">
<h1><span class="header-section-number">1</span> Preamble</h1>
<div id="the-purpose-of-this-site" class="section level2">
<h2><span class="header-section-number">1.1</span> The purpose of this site</h2>
<p>The aim of this site is to introduce cytometry data - specifically mass cytometry data - to data scientists and bioinformaticians not familiar with the technology, the unique challenges relating to the technical and biological origins of the data. The content will be updated periodically as new tools and methods are published.</p>
<p>Along with discussions, we have included code for common processing and analysis tasks for R. Flow cytometry as a scientific discpline has been around for half a century at this point, and a plethora of software and data analysis packages has been created using a multitude of different programming languages. Common to the majority of them, is the heavy focus on the use of The Flow Cytometry Standard (FCS) format. While this format certainly has its strengths, it (and the analysis software developed for it) can seem somewhat convoluted to the uninitiated. Rather than attempting to update these packages for mass cytometry or introduce workarounds, we here focus on extracting and working with the raw expression matrices that most bioinformaticians know from other data types - such as for example transcriptomics (probe and sequence-based alike). The rationale is that once a bioinformatician has an expression matrix and a working understanding of the technical and biological idiosyncracies of the data, most, if not all, common analysis algorithms for expression data can be adapted and applied.</p>
<p>We will also present a selection of common analysis tools. Not with the explicit purpose of applying them in their packaged form, but to explain what goes on under the hood to enable these useful analyses to be applied in a more flexible and customizable way. The main aim here is enabling a high level of autonomy for data scientists working with mass cytometry data by dissecting both the data and common algorithms to its core workable components.</p>
<p>The site consists of a series of code snippets for preparing and analyzing the data. The entire collection of scripts without can be downloaded [here] (<a href="https://raw.githubusercontent.com/biosurf/CyTOF/master/primer.R" class="uri">https://raw.githubusercontent.com/biosurf/CyTOF/master/primer.R</a>). The examples are based on 21 FCS files from healthy control individuals (described below and available from <a href="http://flowrepository.org/experiments/1363/download_ziped_files">flowrepository</a>). Save the files locally and replace the paths in the code below to work with the example data or use your own data.</p>
</div>
<div id="the-example-data-sets" class="section level2">
<h2><span class="header-section-number">1.2</span> The example data sets</h2>
<div id="healthy-donor-samples" class="section level3">
<h3><span class="header-section-number">1.2.1</span> Healthy donor samples</h3>
<p>21 samples of PBMC from healthy control individuals, stained with the <a href="http://www.bio-protocol.org/e1382">Leipold and Maecker 2015 protocol</a>, mixed with EQ normalization beads (Fluidigm Corp), and acquired on a Fluidigm CyTOF Helios mass cytometer using default acquisition parameters. The data exists in two forms: 1) the raw FCS files as generated by the Helios, 2) the samples normalized using the MATLAB method (described in <a href="#beadnorm">section 2.1</a>) - these are files with &quot;normalized&quot; in the file name. Each row in the expression matrix is an event, of which we are only interested in those that represent live, intact, singlet cells (i.e. we want to filter out beads, dead cells, doublets, debris, etc.). Each column contains information about the cells - some are parameters from the data collection by the instrument (e.g. event length, time, etc.), some are markers primarily used for pre-gating (DNA stain, live/dead stain, beads, etc.), some are lineage markers (needed to determine the cell type), and some are activation or cell state makers. <show-table data-table=\"HIMCControlsMarkers\">Here is a table</show-table> with descriptions of each column in the example expression files.</p>
<table>
<colgroup>
<col width="8%" />
<col width="8%" />
<col width="83%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Isotope</th>
<th align="left">Parameter</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Ce140Di</td>
<td align="left">Bead</td>
<td align="left"><strong>Normalization parameter</strong> isotope unique to Fluidigm EQ normalization beads; used to distinguish them from cells during gating</td>
</tr>
<tr class="even">
<td align="left">In115Di</td>
<td align="left">Dead</td>
<td align="left"><strong>Pre-gating parameter</strong> Maleimide-DOTA small molecule as live-dead (cannot efficiently enter cells with undisturbed membranes)</td>
</tr>
<tr class="odd">
<td align="left">Ir191Di</td>
<td align="left">DNA1</td>
<td align="left"><strong>Pre-gating parameter</strong> DNA intercalator (intact cells have high levels)</td>
</tr>
<tr class="even">
<td align="left">Ir193Di</td>
<td align="left">DNA2</td>
<td align="left"><strong>Pre-gating parameter</strong> DNA intercalator (intact cells have high levels)</td>
</tr>
<tr class="odd">
<td align="left">Nd142Di</td>
<td align="left">CD19</td>
<td align="left"><strong>Lineage marker</strong> B cell marker</td>
</tr>
<tr class="even">
<td align="left">Nd146Di</td>
<td align="left">IgD</td>
<td align="left"><strong>Lineage marker</strong> B cell marker</td>
</tr>
<tr class="odd">
<td align="left">Dy164Di</td>
<td align="left">CD20</td>
<td align="left"><strong>Lineage marker</strong> B cell marker</td>
</tr>
<tr class="even">
<td align="left">Nd143Di</td>
<td align="left">CD4</td>
<td align="left"><strong>Lineage marker</strong> T cell marker</td>
</tr>
<tr class="odd">
<td align="left">Nd144Di</td>
<td align="left">CD8</td>
<td align="left"><strong>Lineage marker</strong> T cell marker</td>
</tr>
<tr class="even">
<td align="left">Nd148Di</td>
<td align="left">CD11c</td>
<td align="left"><strong>Lineage marker</strong> used to gate mDCs</td>
</tr>
<tr class="odd">
<td align="left">Sm149Di</td>
<td align="left">CD16</td>
<td align="left"><strong>Lineage marker</strong> used mostly to gate NK cells</td>
</tr>
<tr class="even">
<td align="left">Nd150Di</td>
<td align="left">CD3</td>
<td align="left"><strong>Lineage marker</strong> used to gate T cells</td>
</tr>
<tr class="odd">
<td align="left">Sm154Di</td>
<td align="left">CD14</td>
<td align="left"><strong>Lineage marker</strong> used to gate Monocytes</td>
</tr>
<tr class="even">
<td align="left">Gd155Di</td>
<td align="left">CCR6</td>
<td align="left"><strong>Lineage marker</strong> used to gate Th1/Th2/Th17 cells</td>
</tr>
<tr class="odd">
<td align="left">Gd158Di</td>
<td align="left">CXCR5</td>
<td align="left"><strong>Lineage marker</strong> used to gate T follicular helper cells</td>
</tr>
<tr class="even">
<td align="left">Tb159Di</td>
<td align="left">CXCR3</td>
<td align="left"><strong>Lineage marker</strong> used to gate Th1/Th2/Th17 cells</td>
</tr>
<tr class="odd">
<td align="left">Gd160Di</td>
<td align="left">CCR7</td>
<td align="left"><strong>Lineage marker</strong> used to gate Naive/Central memory/Effector memory/Effector T cells</td>
</tr>
<tr class="even">
<td align="left">Dy162Di</td>
<td align="left">CD45RA</td>
<td align="left"><strong>Lineage marker</strong> used to gate Naive/Central memory/Effector memory/Effector T cells</td>
</tr>
<tr class="odd">
<td align="left">Ho165Di</td>
<td align="left">CD127</td>
<td align="left"><strong>Lineage marker</strong> used to gate Tregs</td>
</tr>
<tr class="even">
<td align="left">Er166Di</td>
<td align="left">CD33</td>
<td align="left"><strong>Lineage marker</strong> used to gate Monocytes</td>
</tr>
<tr class="odd">
<td align="left">Er168Di</td>
<td align="left">CD24</td>
<td align="left"><strong>Lineage marker</strong> used to gate Transitional B cells</td>
</tr>
<tr class="even">
<td align="left">Yb171Di</td>
<td align="left">TCRgd</td>
<td align="left"><strong>Lineage marker</strong> used to gate gamma-delta T cells</td>
</tr>
<tr class="odd">
<td align="left">Yb173Di</td>
<td align="left">CD123</td>
<td align="left"><strong>Lineage marker</strong> used to gate basophils and pDCs</td>
</tr>
<tr class="even">
<td align="left">Yb174Di</td>
<td align="left">CD56</td>
<td align="left"><strong>Lineage marker</strong> used to gate NK cells and NKT cells</td>
</tr>
<tr class="odd">
<td align="left">Yb176Di</td>
<td align="left">CD25</td>
<td align="left"><strong>Lineage marker</strong> used to gate Tregs</td>
</tr>
<tr class="even">
<td align="left">Lu175Di</td>
<td align="left">HLADR</td>
<td align="left"><strong>General protein marker</strong> activation marker on NK and T cells</td>
</tr>
<tr class="odd">
<td align="left">In113Di</td>
<td align="left">CD57</td>
<td align="left"><strong>General protein marker</strong> differentiation marker found on CD8+ T cells and NK cells</td>
</tr>
<tr class="even">
<td align="left">Yb172Di</td>
<td align="left">PD-1</td>
<td align="left"><strong>General protein marker</strong> found on T cells, modulates immune function</td>
</tr>
<tr class="odd">
<td align="left">Tm169Di</td>
<td align="left">ICOS</td>
<td align="left"><strong>General protein marker</strong> found on T cells, modulates immune function</td>
</tr>
<tr class="even">
<td align="left">Er170Di</td>
<td align="left">CD161</td>
<td align="left"><strong>General protein marker</strong> marker of IL17 producing T cells</td>
</tr>
<tr class="odd">
<td align="left">Er167Di</td>
<td align="left">CD28</td>
<td align="left"><strong>General protein marker</strong> found on T cells, necessary for immune activation</td>
</tr>
<tr class="even">
<td align="left">Gd156Di</td>
<td align="left">CD94</td>
<td align="left"><strong>General protein marker</strong> found on CD8+ T cells and NK cells, modulates immune function</td>
</tr>
<tr class="odd">
<td align="left">Gd157Di</td>
<td align="left">CD86</td>
<td align="left"><strong>General protein marker</strong> found on T cells, necessary for immune activation and survival</td>
</tr>
<tr class="even">
<td align="left">Eu151Di</td>
<td align="left">CD38</td>
<td align="left"><strong>General protein marker</strong> found on T cells, B cells, and NK cells, necessary for immune function</td>
</tr>
<tr class="odd">
<td align="left">Sm152Di</td>
<td align="left">CD27</td>
<td align="left"><strong>General protein marker</strong> found on B and T cells, involved in modulation and activation of immune function</td>
</tr>
<tr class="even">
<td align="left">Eu153Di</td>
<td align="left">CD11b</td>
<td align="left"><strong>General protein marker</strong> found on cell of the innate immune system</td>
</tr>
<tr class="odd">
<td align="left">Sm147Di</td>
<td align="left">CD85j</td>
<td align="left"><strong>General protein marker</strong> found on T cells, regulates immune activation</td>
</tr>
<tr class="even">
<td align="left">BCKG190Di</td>
<td align="left">190BCKG</td>
<td align="left"><strong>Background</strong> not used in gating</td>
</tr>
<tr class="odd">
<td align="left">Pb208Di</td>
<td align="left">208Pb</td>
<td align="left"><strong>Background</strong> not used in gating</td>
</tr>
<tr class="even">
<td align="left">Sn120Di</td>
<td align="left">120Sn</td>
<td align="left"><strong>Background</strong> not used in gating</td>
</tr>
<tr class="odd">
<td align="left">I127Di</td>
<td align="left">127I</td>
<td align="left"><strong>Background</strong> not used in gating</td>
</tr>
<tr class="even">
<td align="left">Xe131Di</td>
<td align="left">131Xe</td>
<td align="left"><strong>Background</strong> not used in gating</td>
</tr>
</tbody>
</table>
<p>An R data object with the pre-processed healthy control samples (normalized, pre-gated, transformed - the entire process is detailed below) can be downloaded <a href="https://github.com/biosurf/CyTOF/raw/master/biosurf/cytof/live.Rdata">here</a> if you are mostly interested in post-processing analyses. If so, you may skip to <a href="#analysistasks">section 3</a>.</p>
</div>
<div id="casecontrol-samples" class="section level3">
<h3><span class="header-section-number">1.2.2</span> Case/control samples</h3>
<p>To exemplify analysis of differential abundance, we also use a case/control dataset. We here use the data described by <a href="https://www.nature.com/nbt/journal/v30/n9/full/nbt.2317.html">Bodenmiller et al</a>, which consists of 8 samples of PBMC responding to B-cell receptor (BCR)/Fc receptor (FCR) cross-linking and 8 control PBMC samples. The raw data can be downloaded <a href="https://www.cytobank.org/cytobank/experiments/15713/download_files">here</a> (download the files with &quot;reference&quot; and &quot;BCR-XL&quot; in the filename). These data are pregated and transformed.</p>
</div>
</div>
<div id="installing-the-r-packages-you-will-need" class="section level2">
<h2><span class="header-section-number">1.3</span> Installing the R packages you will need</h2>
<p>We use a variety of packages for the analyses below. In case you are missing some of them, here’s what you will need:</p>
<pre class="r"><code>## From CRAN
install.packages(c(\&quot;scales\&quot;, \&quot;RColorBrewer\&quot;, \&quot;MASS\&quot;, \&quot;Rtsne\&quot;, \&quot;kohonen\&quot;, \&quot;miscTools\&quot;, \&quot;gplots\&quot;, \&quot;Radviz\&quot;, \&quot;igraph\&quot;, \&quot;statmod\&quot;))

## From BioConductor
source(\&quot;https://bioconductor.org/biocLite.R\&quot;)
biocLite(c(\&quot;flowCore\&quot;, \&quot;cytofkit\&quot;, \&quot;ConsensusClusterPlus\&quot;, \&quot;cydar\&quot;, \&quot;flowCL\&quot;, \&quot;CATALYST\&quot;, \&quot;ncdfFlow\&quot;, \&quot;edgeR\&quot;))

## From source
# cytofCore
library(devtools)
install_github(\&quot;nolanlab/cytofCore\&quot;)

# MEM
library(devtools)
download.file(url = \&quot;http://www.nature.com/nmeth/journal/v14/n3/extref/nmeth.4149-S5.zip\&quot;, destfile = \&quot;./MEM.zip\&quot;)
unzip(\&quot;MEM.zip\&quot;)
setwd(\&quot;./MEM\&quot;)
build()
install.packages(\&quot;./\&quot;, repos = NULL, type=\&quot;source\&quot;)
unlink(\&quot;../MEM/\&quot;, recursive = TRUE) # optional cleanup</code></pre>
<p>The R session info is located at the <a href="#sessioninfo">bottom of the page</a>.“,”content“:”# The structure of data from the CyTOF system The following sections include lots of code snippets. The entire raw code can be downloaded <a href="https://raw.githubusercontent.com/biosurf/CyTOF/master/primer.R">here</a>. ## The Flow Cytometry Standard (FCS) format {FCS}</p>
<p>The FCS format is a binary data file standard originally developed for storage of flow cytometry data. As well as the ion count matrix, it has slots for experiment information, channel information, etc. Traditionally, ion counts have been analyzed visually using <a href="https://www.flowjo.com/solutions/flowjo">FlowJo</a> - a commercial software platform for analysis of single-cell cytometry experiments. FlowJo is largely GUI-based and requires a license for both academic and non-academic users. For more automated analysis, the following tools for R can be used to import, manipulate and convert FCS files - or extract the ion counts for downstream analysis in R or other programming environments.</p>
<ul>
<li><a href="http://bioconductor.org/packages/release/bioc/html/flowCore.html">flowCore</a>: facilitates many FCS file related functions. Includes the option to write new FCS files.</li>
<li><a href="https://www.bioconductor.org/packages/release/bioc/html/flowStats.html">flowStats</a>: package for normalizing, gating, and basic and advanced analyses of flow cytometry data.</li>
<li><a href="https://github.com/nolanlab/cytofCore">cytofCore</a>: contains a function to merge FCS files from disrupted CyTOF runs.</li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/CATALYST.html">CATALYST</a>: contains functions for normalization, debarcoding, and quality control.</li>
</ul>
<p>Example of basic FCS file handling:</p>
<pre class="r"><code># Load the flowCore package
library(flowCore)

# Set working directory (directory where you placed the FCS files)
setwd(\&quot;~/cytof_primer/healthy_controls/fcs_files/\&quot;)

# Read an fcs file
fcs_files &lt;- list.files(pattern=&#39;.fcs$&#39;, full=TRUE, ignore.case = TRUE)
fcs &lt;- read.FCS(filename=fcs_files[1], transformation=FALSE)

# Extract expression matrix
exprs &lt;- fcs@exprs

# Explore the channels (columns in the expression matrix)
fcs@parameters@data

# Make colnames human readable using information in the parameter data slot
markers &lt;- gsub(pattern = \&quot;.*_\&quot;, replacement = \&quot;\&quot;, x = as.vector(fcs@parameters@data$desc))
colnames(exprs)[which(!is.na(markers))] &lt;- markers[which(!is.na(markers))]

# Merging fcs files (in case of interrupted runs)
library(cytofCore)
dir.create(\&quot;./combined\&quot;)
cytofCore.concatenateDirectoryFiles(inputDir=\&quot;./\&quot;,outputDir=\&quot;./combined\&quot;,pattern=NULL,overwrite=F,timeParam=\&quot;time\&quot;)</code></pre>
</div>
<div id="parameter-harmonization-harmonize" class="section level2">
<h2><span class="header-section-number">1.4</span> Parameter harmonization {harmonize}</h2>
<p>Note that these channels are <strong>freetext editable</strong> in the Fluidigm CyTOF software. This means that if you are dealing with files from multiple different runs from different instruments and/or different operators (or in some cases even different staining panels), you will need to <strong>harmonize your channel names</strong> to compare samples. The mass information (in the &quot;name&quot; column) is <em>usually</em> the most reliable, but even here errors can occur. When in doubt, consult with the operator before renaming the channels.</p>
</div>
</div>
<div id="data-preprocessing" class="section level1">
<h1><span class="header-section-number">2</span> Data preprocessing</h1>
<p>Preprocessing is essential in any mass cytometry analysis process. Key steps in preprocessing include: <strong>bead normalization</strong>, in which one adjusts for varying instrument performance over time; <strong>pre-gating</strong>, in which live intact singlet cells are obtained; <strong>compensation</strong>, which (if applicable) is used to correct for spillover between the signals of different heavy metal isotopes, and <strong>debarcoding</strong>, which (if applicable) is used to separate different samples in a multiplexed experiment.</p>
<p>For convenience, we here define the different types of channels (columns in the data matrix) used for different purposes:</p>
<pre class="r"><code>pregating_channels &lt;- c(\&quot;Bead\&quot;, \&quot;DNA1\&quot;, \&quot;DNA2\&quot;, \&quot;Dead\&quot;, \&quot;Event_length\&quot;)
lineage_channels &lt;- c(\&quot;CD57\&quot;, \&quot;CD19\&quot;, \&quot;CD4\&quot;, \&quot;CD8\&quot;, \&quot;IgD\&quot;, \&quot;CD11c\&quot;, \&quot;CD16\&quot;, \&quot;CD3\&quot;, \&quot;CD38\&quot;, \&quot;CD27\&quot;, \&quot;CD14\&quot;, \&quot;CXCR5\&quot;, \&quot;CCR7\&quot;, \&quot;CD45RA\&quot;, \&quot;CD20\&quot;, \&quot;CD127\&quot;, \&quot;CD33\&quot;, \&quot;CD28\&quot;, \&quot;CD161\&quot;, \&quot;TCRgd\&quot;, \&quot;CD123\&quot;, \&quot;CD56\&quot;, \&quot;HLADR\&quot;, \&quot;CD25\&quot;)
instrument_channels &lt;- c(\&quot;Time\&quot;, \&quot;Event_length\&quot;, \&quot;Center\&quot;, \&quot;Offset\&quot;, \&quot;Width\&quot;, \&quot;Residual\&quot;, \&quot;sample\&quot;)</code></pre>
<div id="bead-normalization-beadnorm" class="section level2">
<h2><span class="header-section-number">2.1</span> Bead Normalization {beadnorm}</h2>
<p>Because CyTOF instrument performance may vary over time (both within a single run, but more prominently between different runs), it is important to normalize data in a way that limits the impact of the technical variation. The most commonly employed methods are the Fluidigm normalization (DVS Sciences, 2013) and MATLAB normalization (<a href="https://doi.org/10.1002/cyto.a.22271">Finck et al., 2013</a>). Both methods are based on adding beads carrying different heavy metal isotopes to the biological sample, and relying on the beads to leave an extractable signature that can be used to correct signal fluctuations. The Fluidigm normalization method uses a bead &quot;passport&quot; to normalize data to a global standard, while the MATLAB method normalizes using median bead intensities calculated across given experimental data files (e.g. files from different runs on the same machine).</p>
<div id="fluidigm-bead-normalization" class="section level3">
<h3><span class="header-section-number">2.1.1</span> Fluidigm bead normalization</h3>
<p>The Fluidigm bead normalization method is based on the calculation of the mean isotope counts of the beads during production. This global bead profile, called the “Bead Passport”, is unique to each lot of beads synthesized and cannot be adjusted by the user. The first step in the normalization process is for the software to distinguish the beads from the cell sample in question. This is performed using all the major isotopes of the beads (Ce140, Eu151, Eu153, Ho165, and Lu175 for the current Fluidigm EQ beads). The normalization factor is calculated as the ratio between the Bead Passport and the measured counts for each of the beads. The bead channels span a large range of the mass spectrum, and using linear interpolation to masses between or beyond the bead masses allows for the calculation of normalization factors that are then multiplied to each remaining mass channel distributed across the course of the run using the time stamp. This allows for correction of decline in instrument sensitivity during the course of running a sample, and also normalization across samples (see figure below). However, if the instrument used to acquire a sample differs from the Bead Passport in overall signal intensity or in sensitivity across the mass range, the Fluidigm normalization will “force” the sample data to conform to the shape and signal of the Passport (see Figure 4, especially CyTOFv1 red line).</p>
</div>
<div id="matlab-bead-normalization-finck-et-al." class="section level3">
<h3><span class="header-section-number">2.1.2</span> MATLAB bead normalization (Finck et al.)</h3>
<p>The MATLAB implemented normalization algorithm created by <a href="https://doi.org/10.1002/cyto.a.22271">Finck et al. (2013)</a> works by modelling the sample specific drift in count reporting. Briefly, bead singlets are manually identified by biaxial gating of each of the five bead channels against a DNA channel. The intensities of each bead channel over the course of a sample run, are smoothed by calculating local medians in a 500 bead event sliding window. Then, a baseline is calculated by taking the mean of the smoothed intensities for each bead channel. For each bead channel, for each time point, the slope of the linear regression from the channel origin to the given time point is calculated, such that the residual sum of squares between the smoothed bead intensities and the baseline is minimized. For each recorded event, the intensity is multiplied by the slope of corresponding the time point at which is was recorded. Since the MATLAB normalization compares each file to the other files rather than to an externally-determined set of values, there is typically less “forcing” or reshaping of the data than in Fluidigm normalization (see figure below).</p>
<div class="figure">
<img src="https://i.imgur.com/ATkadu0.png" alt="The effects of normalization on Raw data" />
<p class="caption">The effects of normalization on Raw data</p>
</div>
<p>The effects of normalization on Raw data is shown above. Replicate samples were stained, then pooled and split into individual aliquots before freezing in FBS+10% DMSO at -80°C as Takahashi et al. On each day, the sample was thawed on ice, washed 3x in MilliQ water, then resuspended in MilliQ water containing Fluidigm EQ beads and acquired on the indicated CyTOF instrument model. The CyTOFv2 was later upgraded to the v2-to-Helios instrument and another aliquot was run. During data analysis, the EQ beads were gated separately from the cells. The linear y-axis represents the Median Di Dual signal intensity of the indicated mass channel: the Bead data is shown in the top row, while the Cell data is shown on the bottom row (marker on indicated cell population). While the Beads and Cells cover a similar mass range, the signal intensity of the Cells on a specific instrument changes depending on the normalization method (Raw vs. Fluidigm-ver2 vs. MATLAB).</p>
<p>It should be noted that although the bead normalization is useful for adjusting for signal changes due primarily to the instrument performance, it does have some limitations. First, there may be some decrease in signal intensity in the sample that is not reflected in the bead signal. For example, while proper fixation does generally allow cells to withstand the osmotic force of MilliQ water, users have reported decreases in signal intensity during acquisition of large volume barcoded samples. Since the beads are not affected by lengthy suspension in water, the bead signals would not change in the same way that cell signals may. Generally, samples fixed with paraformaldehyde seem more robust than those fixed only with commercial fix/perm solutions. Additionally, samples are typically more stable as pellets. Therefore, this sample signal decrease can often be limited by breaking one large sample into aliquots that would each make 3-5 mL of diluted sample, and only resuspending the pellet of each aliquot shortly before collection. Second, since the beads are not stained with the assay reagents, they are seldom sufficient for true plate-to-plate batch correction, particularly over long periods of time when reagent lots could change.</p>
</div>
<div id="performing-bead-normalization" class="section level3">
<h3><span class="header-section-number">2.1.3</span> Performing bead normalization</h3>
<p>Because the MATLAB method is more commonly used in practice, it is the focus on this wiki. The MATLAB normalization method is implemented with a graphical user interface (GUI), in which the user manually gates the beads using five different bi-axial plots in which each bead metal is plotted against the iridium channel (that measures DNA). This can be installed from <a href="https://github.com/nolanlab/bead-normalization/wiki/Installing-the-Normalizer">here</a>.</p>
<p>While wet lab scientists are accustomed to working with manual gating and GUIs, such as the one in the MATLAB normalization program, this is not the preferred method for data scientists such as bioinformaticians, who commonly wish to analyze a large number of samples, making manual gating very cumbersome. As a consequence, we implemented the MATLAB normalization algorithm in R, such that it may be called from the command-line and run completely without user interference. This process was carried out by translating the MATLAB source code (available from <a href="https://github.com/nolanlab/bead-normalization">here</a>) and making small alterations. The R version can be downloaded from our <a href="https://github.com/biosurf/CyTOF/blob/master/MATLAB_normalization.R">Github repository</a>.</p>
<p>The biggest change from the original version is the fact that bead gating is carried out automatically in the R version using a Support Vector Machine (SVM), which was trained for this specific purpose, using the signals of five bead channels and the DNA channel for a total of seven data sets, for which beads had been gated manually by an experienced wet-lab scientist. The SVM was trained in R using svm from the <a href="https://cran.r-project.org/web/packages/e1071/">e1071 package</a>. The training data was seven samples, each comprising of 250,000 events, of which 4,068-8,715 were defined as beads. The data for each metal channel had been transformed using the formula ArcSinh(data)/c, in which c = 2 for CyTOF Helios data, and furthermore, each metal channel was normalized to values between 0 and 1 for the SVM training, to ensure maximal comparability between different samples. Consequently, new data to be classified using the developed SVM model should be transformed and normalized before &quot;gating&quot; - this is of course incorporated in the developed script. In order to evaluate the performance, a single-layer cross-validation setup was used, because a preliminary run on a smaller subset using a “out-of-the-box” SVM with a radial kernel, turned out to work satisfactory (mean accuracy of 99.997 %, with a mean precision of 99.920 %, and a mean recall of 99.971 %). The training was carried out using a leave-one-out method, in which six data sets was used for training and one data set was used for testing in each round. This should allow for an unbiased performance measure. Based on the obtained results the model was deemed acceptable, and all data sets were used to train a final model that can be used for future bead-normalization tasks - the developed R script requires this <a href="https://github.com/biosurf/CyTOF/blob/master/Bead_SVM_model.Rdata">final model</a> to run. Because the seven training samples for the SVM were all from the CyTOF Helios platform, and they all used EQ beads with the five bead isotopes, <sup>140</sup>Ce, <sup>151</sup>Eu, <sup>153</sup>Eu, <sup>165</sup>Ho, and <sup>175</sup>Lu, the developed method in R only works for data obtained using the EQ beads on a Helios instrument. However, it could potentially easily be extended to handle other data types - it merely requires some training data to develop the proper SVM for gating the given data set type.</p>
</div>
</div>
<div id="data-transformations-transformations" class="section level2">
<h2><span class="header-section-number">2.2</span> Data transformations {transformations}</h2>
<div id="dealing-with-randomized-values" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Dealing with &quot;randomized&quot; values</h3>
<p>When an FCS file is exported by the Helios, non-zero values are randomized negatively and uniformly (]x-1;x]). This means that a count of, for example, 9 will be evenly distributed between 8 and 9 for each channel, for each event. This is mostly done for historical reasons, as visualizing count data from the CyTOF produce different plots than intensity data (from flow cytometers). Randomization is done (by default) during conversion of IMD to FCS by the Fluidigm software. Randomized counts can be converted back to the original value by rounding every decimal up to the nearest whole number.</p>
<pre class="r"><code># Check if counts are randomized (expression matrix should only contain integers)
is.wholenumber &lt;- function(x, tol = .Machine$double.eps^0.5)  abs(x - round(x)) &lt; tol
table(is.wholenumber(exprs[,c(pregating_channels, lineage_channels)]))

# You may revert back to original counts (only applicable to measured channels - i.e. exclude time, event length, etc.)
exprs &lt;- cbind(ceiling(exprs[,c(pregating_channels, lineage_channels)]), exprs[,instrument_channels])</code></pre>
<p>For multidimensional analysis this is definitely recommendable not to add unnecessary noise to the data, but for the purpose of the 2D visualizations we are doing for pre-gating, we will leave the randomized counts for now.</p>
</div>
<div id="arcsinh-transformation" class="section level3">
<h3><span class="header-section-number">2.2.2</span> ArcSinh transformation</h3>
<p>Commonly, mass cytometry ion counts will be ArcSinh transformed - this transformation will retain linearity in the low end of the count spectrum, but it resembles a log transformation in the high end. If counts are divided by a co-factor before the transformation, the range of ion counts for which linearity is retained may be adjusted. For CyTOF data, a co-factor of 5 is typically used. Transformations are done primarily for visualization purposes and co-factors are empirically determined. In some cases, adjusting co-factors slightly may improve visualizations.</p>
<p>Continuing from the above code snippet for loading an FCS, the data may be ArcSinh-transformed as follows:</p>
<pre class="r"><code>library(MASS)
library(RColorBrewer)
k &lt;- 11; my.cols &lt;- rev(brewer.pal(k, \&quot;RdYlBu\&quot;))

# Set co-factor
cofac &lt;- 5

# Make arcsinh transformed expression matrix (with the exception of time and event_length, which should remain linear)
exprs_trans &lt;- cbind(asinh(exprs[,c(pregating_channels, lineage_channels)]/cofac), exprs[,instrument_channels])

# Plot two-marker example of transformation effect
plot(exprs[,c(\&quot;CD8\&quot;, \&quot;CD4\&quot;)], pch=\&quot;.\&quot;, col=\&quot;grey\&quot;, main=\&quot;CD8 vs CD4\&quot;)
z &lt;- kde2d(exprs[,\&quot;CD8\&quot;], exprs[,\&quot;CD4\&quot;], n=50)
contour(z, drawlabels=FALSE, nlevels=k, col=my.cols, add=TRUE)

plot(exprs_trans[,c(\&quot;CD8\&quot;, \&quot;CD4\&quot;)], pch=\&quot;.\&quot;, col=\&quot;grey\&quot;, main=\&quot;CD8 vs CD4 (transformed counts)\&quot;)
z &lt;- kde2d(exprs_trans[,\&quot;CD8\&quot;], exprs_trans[,\&quot;CD4\&quot;], n=50)
contour(z, drawlabels=FALSE, nlevels=k, col=my.cols, add=TRUE)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/OmhBhh8.png" alt="Raw ion counts vs arcsinh transformed counts" />
<p class="caption">Raw ion counts vs arcsinh transformed counts</p>
</div>
<p>Both the classic Logicle transformation and ArcSinh works by compressing values in the upper end of the spectrum and enhancing the resolution on the lower end:</p>
<div class="figure">
<img src="https://i.imgur.com/t2XTC1P.png" alt="Logicle vs arcsinh transformation" />
<p class="caption">Logicle vs arcsinh transformation</p>
</div>
<p>When plotting the variation for each marker before and after transformation, the difference is quite clear:</p>
<div class="figure">
<img src="https://i.imgur.com/oWAJwoG.png" alt="The effect of transformation on data distribution" />
<p class="caption">The effect of transformation on data distribution</p>
</div>
</div>
</div>
<div id="pre-gating" class="section level2">
<h2><span class="header-section-number">2.3</span> Pre-gating</h2>
<p>Pre-gating has the purpose to filter the data to consist only of live, intact, single cells. Traditionally, this is carried out in four steps, carried out for each sample. There are number of ways one can semi or fully automate this process (using unsupervised or supervised machine learning for instance), but for clarity we will perform &quot;old school&quot; pre-gating below - sped up by pre-gating all samples in one go (insofar as samples are comparable to a degree that enables this). In the following sections, we will work on all the samples as a single object, so first we reload the data into a data.frame.</p>
<pre class="r"><code># For the next step, set working directory to the normalized samples
setwd(\&quot;~/cytof_primer/fcs_files/normed/\&quot;)

# Read in your normalized files and concatenate in a data frame (in this example we subset each matrix to 10,000 random events)
fcs_files &lt;- list.files(pattern=&#39;.fcs$&#39;, full=TRUE, ignore.case = TRUE)
exprs_set &lt;- data.frame()
sample &lt;- c()
for(i in fcs_files) {
  fcs &lt;- read.FCS(filename=i, transformation=FALSE) 
  exprs &lt;- fcs@exprs
  markers &lt;- gsub(pattern = \&quot;.*_\&quot;, replacement = \&quot;\&quot;, x = as.vector(fcs@parameters@data$desc))
  colnames(exprs)[which(!is.na(markers))] &lt;- markers[which(!is.na(markers))]
  # exprs &lt;- cbind(floor(exprs[,!colnames(exprs) %in% instrument_channels]), exprs[,colnames(exprs) %in% instrument_channels])
  exprs_set &lt;- rbind(exprs_set, exprs[sample(nrow(exprs), 10000),])
  sample &lt;- append(sample, rep(i, 10000))
}
exprs_set$sample &lt;- sample
instrument_channels &lt;- append(instrument_channels, \&quot;sample\&quot;)

# Make arcsinh transformed expression matrix (excluding channels that should remain linear - e.g. time, event length, etc.)
cofac &lt;- 5 # For CyTOF data, the co-factor is set to 5
exprs_set_trans &lt;-cbind(asinh(exprs_set[,c(pregating_channels, lineage_channels)]/cofac), exprs_set[,instrument_channels])</code></pre>
<div id="batch-correction" class="section level3">
<h3><span class="header-section-number">2.3.1</span> Batch correction</h3>
<p>Insofar as all samples are run on the same instrument, with the same panel, by the same operator, and bead normalized together, it should be possible to avoid pre-gating all your samples one at the time, and rather doing it in one go. You may check that the distributions of bead, DNA, and live/dead measurements are somewhat similar:</p>
<pre class="r"><code>library(scales)
library(RColorBrewer)

cols &lt;- brewer.pal(5,&#39;Set1&#39;)
par(mfrow=c(1,5))

for(c in 1:length(pregating_channels)) {
  plot(1, type = \&quot;n\&quot;, xlim = c(0, max(exprs_set_trans[,pregating_channels[c]])), ylim = c(1, length(fcs_files) + 2), axes = FALSE, xlab = \&quot;\&quot;, ylab = \&quot;\&quot;, main=pregating_channels[c], cex.main=3)
  for(s in 1:length(fcs_files)){
    dens &lt;- density(exprs_set_trans[exprs_set_trans$sample==fcs_files[s],pregating_channels[c]])
    dens$y = dens$y+(length(fcs_files)-s+1)
    par(bg=NA)
    lines(dens)
    polygon(dens, col=alpha(cols[c],0.4), border = NA)
    abline(h =(length(fcs_files)-s+1), lwd = 0.5)
  }
}</code></pre>
<p><img src=\"https://i.imgur.com/rRo1Ycg.png\" alt=\"Visualization of the manual pre-gating strategy\" style=\"width: 800px;\"/></p>
<p>As can be seen all five pre-gating channels are largely comparable in all samples, and batch correction is not necesarry for pre-gating.</p>
<p>In case your data doesn’t look quite as nice (these samples were run on the same instrument, by the same operator, during the course of a single day), you can correct for batch effect channel-by-channel using a range based approach (basically a linear alignment of densities) or a warping normalization that enables non-linear adjustments to individual peaks in the data (frequently referred to as landmarks).</p>
<p>First, let’s construct an example:</p>
<pre class="r"><code># Create two toy batches of bimodal densities (the cydar packages only takes lists of lists as input)
batches &lt;- list()
x &lt;- list()
ex &lt;- matrix(c(rnorm(500,2,1), rnorm(500,25,2)), ncol=1)
colnames(ex) &lt;- \&quot;CDx\&quot;
x[[\&quot;sample_Y\&quot;]] &lt;- ex
batches[[\&quot;Batch1\&quot;]] &lt;- x
ex &lt;- matrix(c(rnorm(500,2.1,1.1), rnorm(500,15,2)), ncol=1)
colnames(ex) &lt;- \&quot;CDx\&quot;
x[[\&quot;sample_Y\&quot;]] &lt;- ex
batches[[\&quot;Batch2\&quot;]] &lt;- x

# Plot the marker distributions
plot(density(batches$Batch1$sample_Y), main=\&quot;uncorrected samples\&quot;, ylim=c(0,0.09))
points(density(batches$Batch2$sample_Y), type=\&quot;l\&quot;, col=\&quot;red\&quot;)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/3SdFPxL.png" alt="Uncorrected samples" />
<p class="caption">Uncorrected samples</p>
</div>
<p>The distributions are clearly unaligned and we will need to correct for this before we can analyze them together:</p>
<pre class="r"><code># Run the normalization
library(cydar)
corrected &lt;- normalizeBatch(batches, batch.comp = NULL, mode=\&quot;warp\&quot;)

# Plot the results
plot(density(corrected$Batch1$sample_Y), main=\&quot;batch corrected samples\&quot;)
points(density(corrected$Batch2$sample_Y), type=\&quot;l\&quot;, col=\&quot;red\&quot;)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/NrlRQJV.png" alt="Batch corrected data" />
<p class="caption">Batch corrected data</p>
</div>
<p>Voila.</p>
<p>Channel-by-channel correction should be performed with care. A common source of noise is the stability (or lack thereof) of the antibody cocktail. In practice, this means that in one sample, a given channel can be virtually noise-free, whereas in another, noise can be introduced into that very channel - even if the same antibody cocktail is used. When warping such two channels to be comparable without a priori knowledge about what is noise and what is signal, one is basically making the noise comparable by &quot;contaminating&quot; the noise free sample.</p>
<p>Tools for system-wide batch correction are available. For example, the <a href="https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btx196">distribution-matching residual networks</a> approach. However, the code for this tool is not yet available in an implementation-ready form. Python scripts are found <a href="https://github.com/ushaham/BatchEffectRemoval">here</a>, but issues like the example data being hard coded as the input data in the main script, prevents easy implementation of this otherwise interesting method.</p>
<p>If none of this does help on your data (this is obviously an oversimplified example), one solution is to analyze your samples individually and then compare the results. Batch correcting mass cytometry data is definitely non-trivial, and a method for modeling and removing system-wide as well as channel-specific noise has yet to see the light of day.</p>
</div>
<div id="gating-for-cells-beads-vs.dna" class="section level3">
<h3><span class="header-section-number">2.3.2</span> Gating for cells (beads vs. DNA)</h3>
<p>The first step is to gate out beads, cell/bead dobulets, and debris. This is done by generating a biaxial plot with a DNA channel (e.g. <sup>191</sup>Ir) on one axis and a bead-only isotope (<sup>140</sup>Ce) on the other. A single bead should be DNA-negative and <sup>140</sup>Ce-positive, and cells should be DNA-positive and <sup>140</sup>Ce-negative. Double negatives are considered to be debris and double positives are commonly doublets. Below is a code snippet for conducting this step after initializing as described <a href="#fcs">above</a>.</p>
<pre class="r"><code># Load necessary packages
library(MASS)
library(RColorBrewer)

# Make color palette using color brewer
k &lt;- 11; my.cols &lt;- rev(brewer.pal(k, \&quot;RdYlBu\&quot;))

# Gate for cells (beads vs. DNA)
sub &lt;- exprs_set_trans[exprs_set_trans$sample==fcs_files[4],c(\&quot;Bead\&quot;, \&quot;DNA1\&quot;)]
z &lt;- kde2d(x=sub[,1], y=sub[,2], n=50, h=max(bandwidth.nrd(sub[,1]), bandwidth.nrd(sub[,2])))

# Generating a plot for gating visualization
plot(sub, pch=\&quot;.\&quot;, cex=0.1, col=\&quot;gray\&quot;, xlab=colnames(sub)[1], ylab=colnames(sub)[2], xlim=c(0,max(sub[,1])), ylim=c(0,max(sub[,2])))
contour(z, drawlabels=FALSE, nlevels=k, col=my.cols, add=TRUE)

# Set gate boundaries - the values should be manually adjusted
left &lt;- 0; right &lt;- 3; lower &lt;- 6; upper &lt;- 9
rect(left,lower,right,upper,border=\&quot;red\&quot;)
cells &lt;- exprs_set_trans[exprs_set_trans[,\&quot;Bead\&quot;] &lt; right,]; cells &lt;- cells[cells[,\&quot;DNA1\&quot;] &gt; lower &amp; cells[,\&quot;DNA1\&quot;] &lt; upper,]
paste(\&quot;cells: \&quot;, round((nrow(cells)/nrow(exprs_set_trans))*100, digits=2), \&quot;%\&quot;, sep=\&quot;\&quot;)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/SHUUXb7.png" alt="Gating for cell" />
<p class="caption">Gating for cell</p>
</div>
</div>
<div id="gating-for-intact-cells-dna1-vs-dna2" class="section level3">
<h3><span class="header-section-number">2.3.3</span> Gating for intact cells (DNA1 vs DNA2)</h3>
<p>The second step is to gate for intact cells - usually by generation a bixaial plot with the two DNA channels (<sup>191</sup>Ir and <sup>193</sup>Ir) and gating for events, which express both at a balanced level. If events express only little DNA it is likely debris, but if DNA expression is too high, they are likely cell doublets.</p>
<pre class="r"><code># Gate for intact cells (DNA1 vs DNA2)
sub &lt;- exprs_set_trans[,c(\&quot;DNA1\&quot;, \&quot;DNA2\&quot;)]
z &lt;- kde2d(sub[,1], sub[,2], n=50)

# Generating a plot for gating visualization
plot(sub, pch=\&quot;.\&quot;, cex=0.1, col=\&quot;gray\&quot;, xlab=colnames(sub)[1], ylab=colnames(sub)[2], xlim=c(0,max(sub[,1])), ylim=c(0,max(sub[,2])))
contour(z, drawlabels=FALSE, nlevels=k, col=my.cols, add=TRUE)

# Set gate boundaries - the values should be manually adjusted
left &lt;- 6.3; right &lt;- 8.2; lower &lt;- 6.8; upper &lt;- 8.85
rect(left,lower,right,upper,border=\&quot;red\&quot;)
intact &lt;- cells[cells[,\&quot;DNA1\&quot;] &gt; left &amp; cells[,\&quot;DNA1\&quot;] &lt; right,]; intact &lt;- intact[intact[,\&quot;DNA2\&quot;] &gt; lower &amp; intact[,\&quot;DNA2\&quot;] &lt; upper,]
paste(\&quot;intact cells: \&quot;, round((nrow(intact)/nrow(cells))*100, digits=2), \&quot;%\&quot;, sep=\&quot;\&quot;)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/rpi3Lh2.png" alt="Gating for intact cells" />
<p class="caption">Gating for intact cells</p>
</div>
</div>
<div id="gating-for-singlets-event-length-vs-dna" class="section level3">
<h3><span class="header-section-number">2.3.4</span> Gating for singlets (event length vs DNA)</h3>
<p>The third step is to gate out doublets more specifically. This can be done by biaxially plotting event length against a DNA channel and gating for the events which have a length within the range of most events.</p>
<pre class="r"><code># Gate for singlets (event length vs. DNA)
sub &lt;- intact[,c(\&quot;Event_length\&quot;, \&quot;DNA1\&quot;)]
z &lt;- kde2d(sub[,1], sub[,2], n=50)

# Generating a plot for gating visualization
plot(sub, pch=\&quot;.\&quot;, cex=0.1, col=\&quot;gray\&quot;, xlab=colnames(sub)[1], ylab=colnames(sub)[2], xlim=c(0,max(sub[,1])), ylim=c(0,max(sub[,2])))
contour(z, drawlabels=FALSE, nlevels=k, col=my.cols, add=TRUE)

# Set gate boundaries - the values should be manually adjusted
left &lt;- 11; bottom &lt;- min(sub[,2]); right &lt;- 21; upper &lt;- max(sub[,2])
rect(left,bottom,right,upper,border=\&quot;red\&quot;, lwd=3) # Draw cells gate (format: ll x,y; ur x,y)
singlets &lt;- intact[intact[,\&quot;Event_length\&quot;] &gt; left &amp; intact[,\&quot;Event_length\&quot;] &lt; right,]
paste(\&quot;Intact singlets: \&quot;, round((nrow(singlets)/nrow(intact))*100, digits=2), \&quot;%\&quot;, sep=\&quot;\&quot;)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/32Irz1E.png" alt="Gating for singlets" />
<p class="caption">Gating for singlets</p>
</div>
</div>
<div id="gating-for-live-cells-livedead-stain-vs-dna" class="section level3">
<h3><span class="header-section-number">2.3.5</span> Gating for live cells (live/dead stain vs DNA)</h3>
<p>The final step is to remove dead cells. This is typically carried out by using a viability stain. This staining takes place before ligation of antibody-probes. Because dead cells have disrupted membranes, the stain can enter them and form bonds with intracellular molecules. The gating itself is carried out by biaxially plotting the viability stain and a DNA channel and gating out cells with a high amount of staining.</p>
<pre class="r"><code># Gate for live
sub &lt;- singlets[,c(\&quot;Dead\&quot;, \&quot;DNA1\&quot;)]
z &lt;- kde2d(sub[,1], sub[,2], n=50)

# Generating a plot for gating visualization
plot(sub, pch=\&quot;.\&quot;, cex=0.1, col=\&quot;gray\&quot;, xlab=colnames(sub)[1], ylab=colnames(sub)[2], xlim=c(0,max(sub[,1])), ylim=c(0,max(sub[,2])))
contour(z, drawlabels=FALSE, nlevels=k, col=my.cols, add=TRUE)

# Set gate boundaries - the values should be manually adjusted
left &lt;- 0; bottom &lt;- min(sub[,2]); right &lt;- 5; upper &lt;- max(sub[,2])
rect(left,bottom,right,upper,border=\&quot;red\&quot;, lwd=3)
live &lt;- singlets[singlets[,\&quot;Dead\&quot;] &lt; right,]
paste(\&quot;Live intact singlets: \&quot;, round((nrow(live)/nrow(singlets))*100, digits=2), \&quot;%\&quot;, sep=\&quot;\&quot;)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/UyfMgo9.png" alt="Gating for live cells" />
<p class="caption">Gating for live cells</p>
</div>
<p>Finally, you can print the recovery percentage for each sample and save the live intact singlets to a new FCS file if you need to.</p>
<pre class="r"><code>for (i in fcs_files) {
  print(paste(\&quot;Total recovery for \&quot;, i, \&quot;: \&quot;, round((nrow(live[live$sample==i,])/nrow(exprs_set_trans[exprs_set_trans$sample==i,]))*100, digits=2), \&quot;%\&quot;, sep=\&quot;\&quot;))
}

# Save FCS containing live intact singlets
ff2 &lt;- flowFrame(xt_live_intact_singlets, ff@parameters, ff@description)
suppressWarnings(write.FCS(ff2, filename = paste(\&quot;./\&quot;,  file, \&quot;live_intact_singlets\&quot;, sep=\&quot;\&quot;), what=\&quot;numeric\&quot;, delimiter = \&quot;\\\\\&quot;))</code></pre>
<p>More automated options for pre-gating can be conceived. One option is to train a supervised model, such as a support vector machine, on a sample for which manual gating has already been carried out. However, in this case it is also necessary to normalize and batch correct the data for alignment of the distributions of the pre-gating channels.</p>
</div>
</div>
<div id="compensation" class="section level2">
<h2><span class="header-section-number">2.4</span> Compensation</h2>
<p>Compensation is a method for correcting signal overlap, or &quot;spillover&quot;, between the different channels used in cytomtetry experiments. While being an established method in flow cytometry, a recent study (<a href="http://dx.doi.org/10.1016/j.cels.2018.02.010">Chevrier et al., 2018</a>) highlight that mass cytometry experiments are also affected by signal spillover. This leads to noisy expression matrices, and as a consequence, compensation must be applied.</p>
<p>To perform compensation, you need to directly quantify the overlap between isotopes using beads (or cells) stained with a single isotope. Using these single-stained beads it is possible to estimate a spillover matrix from the populations identified as belonging to each isotope, and then one can apply the inverse of the matrix to compensate the data for CyTOF runs through multiplication.</p>
<p>The <a href="http://bioconductor.org/packages/release/bioc/html/CATALYST.html">CATALYST tool</a> offers an implementation of a compensation method, which views spillover as linear (traditional approach) and uses non-negative linear least squares to avoid yielding negative expression values. Additionally, their vignette is quite extensive and helpful.</p>
<p>But how can you tell if your data needs compensation? In reality this may be hard, but one approach could be to use a biaxial plot to look for unexpected correlations between channels in your dataset. E.g. you would <em>usually</em> not expect a correlation between a barcoding channel and a surface marker, but you may expect a correlation between two maturity markers or two activation markers. So if you see something like the leftmost plot below (which shows a barcoding vs. a marker channel), you have a strong indication that something is wrong. Fortunately, it can be corrected using CATALYST to yield the rightmost plot. For this dataset (which is one of our own), none of the single-stained beads were actually stained with <sup>110</sup>Pd, and consequently any signal detected in this channel must be spillover.</p>
<div class="figure">
<img src="https://i.imgur.com/a6FzOTC.png" alt="Compensation example. Left: Uncompensated data with a large spillover from the Cd114 channel to the Pd110 channel. Right: Compensated data." />
<p class="caption">Compensation example. Left: Uncompensated data with a large spillover from the Cd114 channel to the Pd110 channel. Right: Compensated data.</p>
</div>
</div>
<div id="debarcoding" class="section level2">
<h2><span class="header-section-number">2.5</span> Debarcoding</h2>
<p>A GUI implementation of the single cell debarcoder may be installed from <a href="https://github.com/ParkerICI/premessa">here</a> or <a href="http://bioconductor.org/packages/release/bioc/html/CATALYST.html">here</a>, but the CATALYST implementation may also be used without the GUI. However, it is still advisable to adjust the barcode separation cut-offs manually, and for this purpose it is necessary to plot the data. The original debarcoding paper was published by <a href="http://dx.doi.org/10.1038/nprot.2015.020">Zunder et al. (2013)</a>, and the basic principle is that you use combinations of palladium isotopes to label individual samples, which you can then pool and run together on the CyTOF machine to avoid technical variation. To separate the different samples, one must identify events for which the difference in the signal between the present and absent palladium channels is clear.</p>
</div>
</div>
<div id="data-analysis" class="section level1">
<h1><span class="header-section-number">3</span> Data analysis</h1>
<div id="dimensionality-reduction-dimred" class="section level2">
<h2><span class="header-section-number">3.1</span> Dimensionality reduction {dimred}</h2>
<div id="principal-component-analysis-pca" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Principal component analysis {pca}</h3>
<p>The most commonly used algorithm for dimensionality reduction in the biological sciences is principal component analysis (PCA). For mass cytometry data, a PCA will often capture 40-50% of the variance on the first 2-3 PCs, but since the distance between events is linear, data existing on a nonlinear manifold will be misrepresented <a href="http://www.jimmunol.org/content/195/3/773.long">(Chester &amp; Maecker 2015)</a>. Whether this is an issue with CyTOF data or not depends heavily on the source of data. For example, as cells in the hematopoietic system develop in a hierarchical structure, the transition between some cells will result in a continuum rather than distinct clusters, which may not be entirely captured by PCA.</p>
<pre class="r"><code>pca &lt;- prcomp(live[live$sample==fcs_files[1],lineage_channels], scale. = TRUE)
plot(pca$x[,1:2], pch=21, bg=alpha(\&quot;slategray4\&quot;,0.6), col=\&quot;grey\&quot;)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/Dld6ETL.png" alt="2D PCA" />
<p class="caption">2D PCA</p>
</div>
</div>
<div id="t-sne-tsne" class="section level3">
<h3><span class="header-section-number">3.1.2</span> t-SNE {tsne}</h3>
<p>The t-Distributed Stochastic Neighbor Embedding (t-SNE) dimensionality reduction algorithm is a popular tool for dimensionality reduction of cytometry data (van der Maaten and Hinton, 2008). The distance between data points are represented by a t-distribution, meaning that multidimensional distances are preserved to a greater degree than PCA. However, there are a number of features of the algorithm that can lead to misinterpretation by the unaware user. For example, t-SNE (intentionally) does not preserve distances nor density, and it only preserves nearest-neighbors if reasonable hyperparameters are selected. This means that the sizes of the clusters are not always indicative of the variability within the cluster, and similarly, the distance between two clusters is not always proportional to how different they really are. In this regard, the algorithm is quite sensitive to the user making an informed choice of the perplexity and epsilon hyperparameters to reasonably represent the underlying structure of the data (See Martin Wattenberg et al.’s excellent blog post on <a href="http://distill.pub/2016/misread-tsne/">how to use t-SNE efficiently</a>.). Another feature that may be considered a disadvantage to some is that t-SNE depends on a random start and is non-deterministic even on the same data unless the same seed and iteration limit is used for each run. Additionally, removing or adding even a few data points from the input set will affect the layout of the remaining events, and consequently it is not directly possible to add samples to an existing t-SNE plot. Lastly, t-SNE does not allow for identical events, which, in the case of immune profiling where one would expect at least some identical cells in each cluster, means discarding or scrambling data. Dimensionality reduction definitely has a place in mass cytometry analysis pipelines, but for these reasons, it should be limited to exploratory and/or post-analysis visualizations. t-SNE is implemented in a number of languages including R, Python, C++, etc., and as a plug-in to FlowJo X. The MATLAB implementation of t-SNE often used for mass cytometry data in GUI pipelines is referred to as viSNE (<a href="https://doi.org/10.1038/nbt.2594">Amir et al., 2013</a>).</p>
<p>Before running t-SNE, it is necessary to remove identical events (if any) from the sample (duplicate rows).</p>
<pre class="r"><code>table(duplicated(live[live$sample==fcs_files[1],lineage_channels]))
# No duplicates here, but if so, they can be removed:
live &lt;- live[-which(duplicated(live[live$sample==fcs_files[1],lineage_channels]))]</code></pre>
<pre class="r"><code>library(Rtsne)
set.seed(42)
tsne &lt;- Rtsne(live[live$sample==fcs_files[1],lineage_channels])
plot(tsne$Y[,1:2], pch=21, bg=alpha(\&quot;slategray4\&quot;,0.6), col=\&quot;grey\&quot;, xlab=\&quot;\&quot;, ylab=\&quot;\&quot;, lwd=0.5)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/20LPhRO.png" alt="2D t-SNE" />
<p class="caption">2D t-SNE</p>
</div>
<p>Even before clustering, it is possible to get some sense of the cell composition in a sample by selectively coloring cells by marker expression on the t-SNE plot:</p>
<pre class="r"><code>par(mfrow=c(2,2))
for(i in c(\&quot;CD4\&quot;, \&quot;CD8\&quot;, \&quot;CD19\&quot;, \&quot;CD56\&quot;)) {
  cols_ramp &lt;- colorRamp(c(\&quot;slategray4\&quot;, \&quot;red\&quot;))
  marker_exprs_scaled &lt;- (live[live$sample==fcs_files[1],i] - min(live[live$sample==fcs_files[1],i]))/diff(range(live[live$sample==fcs_files[1],i]))
  cols &lt;- rgb(cols_ramp(marker_exprs_scaled), maxColorValue = 256)
  plot(tsne$Y[,1:2], pch=21, bg=alpha(cols,0.6), col=\&quot;grey\&quot;, xlab=\&quot;\&quot;, ylab=\&quot;\&quot;, lwd=0.5, main=i)
}</code></pre>
<div class="figure">
<img src="https://i.imgur.com/6KXyzP4.png" alt="Cells highlighted by marker expression" />
<p class="caption">Cells highlighted by marker expression</p>
</div>
</div>
<div id="umap" class="section level3">
<h3><span class="header-section-number">3.1.3</span> UMAP</h3>
<p>Uniform Manifold Approximation and Projection (UMAP) is a newer addition to the family of CyTOF-applicable dimensionality reduction methods (<a href="http://dx.doi.org/10.1038/nbt.4314">Becht et al., 2018</a>). It works in a fashion that is more similar to t-SNE than PCA, which means that it also views each cell as a member of a neighborhood of similar cells. However, it has several advantages over t-SNE. The first one is speed. For a small dataset there may not be a significant difference, bu you will surely notice it for a million cells. Furthermore, UMAP preserves global distances in the data a lot better than t-SNE. Recall that distances have no meaning in a t-SNE plot - you can not say anything about <em>how</em> different clusters are with t-SNE and distances between clusters tend to be fairly constant. In UMAP plots, distances once again have meaning, although they are not linear and directly interpretable as for PCA.</p>
<p>UMAP, like t-SNE also requires setting a seed for each run to make it reproducible. UMAP is implemented in several languages including R and Python.</p>
<pre class="r"><code>library(umap)
set.seed(42)
umap &lt;- umap(live[live$sample==fcs_files[1],lineage_channels], n_neighbors = 15, min_dist = 0.2, metric = &#39;euclidean&#39;, method = &#39;naive&#39;)
plot(umap$layout[,1:2], col = &#39;grey&#39;, bg = scales::alpha(\&quot;slategray4\&quot;,0.6), pch = 21, xlab = &#39;UMAP1&#39;, ylab = &#39;UMAP2&#39;)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/TTxTrT4.jpg" alt="2D UMAP" />
<p class="caption">2D UMAP</p>
</div>
<p>Again, it is possible to get a pretty good insight into cell composition in a sample by selectively coloring cells by marker expression on the UMAP plot - you will also notice how this is similar to the patterns seen with t-SNE:</p>
<pre class="r"><code>par(mfrow=c(2,2))
for(i in c(\&quot;CD4\&quot;, \&quot;CD8\&quot;, \&quot;CD19\&quot;, \&quot;CD56\&quot;)) {
  cols_ramp &lt;- colorRamp(c(\&quot;slategray4\&quot;, \&quot;red\&quot;))
  marker_exprs_scaled &lt;- (live[live$sample==fcs_files[1],i] - min(live[live$sample==fcs_files[1],i]))/diff(range(live[live$sample==fcs_files[1],i]))
  cols &lt;- rgb(cols_ramp(marker_exprs_scaled), maxColorValue = 256)
  plot(umap$layout[,1:2], pch=21, bg=alpha(cols,0.6), col=\&quot;grey\&quot;, xlab=\&quot;\&quot;, ylab=\&quot;\&quot;, lwd=0.5, main=i)
}</code></pre>
<div class="figure">
<img src="https://i.imgur.com/irRiEXw.jpg" alt="Cells highlighted by marker expression" />
<p class="caption">Cells highlighted by marker expression</p>
</div>
</div>
</div>
<div id="cell-subset-detection-clustering-clustering" class="section level2">
<h2><span class="header-section-number">3.2</span> Cell subset detection (clustering) {clustering}</h2>
<p>Deconvoluting a cell mixture may be a goal in itself, and certainly a necessity for meaningful comparison of samples or cell populations whether transversal or longitudinal. The traditional approach relies on manual gating on biaxial plots, but as the number of features are increased, the manual strategy becomes increasingly laborious. A number of methods are therefore seeking to automate or semi-automate the process - not only in the interest of speed, but unsupervised methods may also reveal cell populations overlooked when following standard gating strategies.</p>
<p>Automated strategies are usually based on some form of clustering. Since the hematopoietic continuum is a cellular hierarchy, the obvious first choice is simply hierarchical clustering. However, this method is by some considered suboptimal, usually ascribed to the need to manually select the number of clusters based on a distance cutoff in the resulting dendrogram. This forces the user to guess the number of populations in the sample: selecting too many will over-separate the data into meaningless small clusters; selecting too few will agglomerate smaller, but potentially interesting clusters into larger ones. Additionally, with a computational complexity of O(n) and a typical yield of ~200.000 cells per sample, this can be a time consuming algorithm to run on CyTOF data. However, multiple tools still utilize hierarchical clustering in their algorithms. SPADE (Qiu et al. 2011), for example, overcomes the computational challenges and the risk of missing rare populations by performing density dependent down-sampling before hierarchical clustering.</p>
<p>Phenograph (automatically decides the optimal number of clusters) and FlowSOM (you need to provide an informed guess of the number of clusters) were recently shown to be the best performing algorithms in a <a href="http://onlinelibrary.wiley.com/doi/10.1002/cyto.a.23030/abstract">comparison of the main algorithms by Weber and Robinson 2016</a>. It is of course worth mentioning that there are many other methods available (eg. <a href="http://www.cellaccense.com/">ACCENSE</a> and <a href="http://bioconductor.org/packages/release/bioc/vignettes/cytofkit/inst/doc/cytofkit_example.html">clusterX</a>). We will also briefly touch upon a few tools such as e.g. X-shift (Samusik et al. 2016), or force-directed layout-based methods, e.g. SCAFFoLD (Spitzer et al. 2015), which currently do not have R command based interfaces, as well as cover tools with integrated clustering (e.g. <a href="https://github.com/nolanlab/spade/wiki">SPADE</a> and <a href="https://github.com/nolanlab/citrus">citrus</a>) in greater depth in later sections.</p>
<div id="phenograph" class="section level3">
<h3><span class="header-section-number">3.2.1</span> Phenograph</h3>
<p>Phenograph is an algorithm which directly assigns single cells to a particular cluster while taking the all measured dimensions into account. Phenograph employs a k-nearest neighbor graph (k-NNG), in which each cell is a node that is connected to its nearest neighbors by edges, to cluster cells into phenotypically similar groups - cellular populations. Clusters consist of nodes with high inter-connectivity. Phenograph is implemented in the R package &quot;cytofkit&quot;.</p>
<p>Phenograph also requires removal of duplicates (see <a href="#tsne">t-SNE section</a> above).</p>
<pre class="r"><code>library(cytofkit)
clusters_pg &lt;- cytof_cluster(xdata = live[live$sample==fcs_files[1],lineage_channels], method = \&quot;Rphenograph\&quot;)</code></pre>
<pre class="r"><code># Visualize results on PCA
getPalette = colorRampPalette(brewer.pal(9, \&quot;Set1\&quot;))
cols &lt;- getPalette(max(clusters_pg))
plot(pca$x[,1:2], pch=21, bg=alpha(col=cols[clusters_pg],0.6), col=\&quot;grey\&quot;)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/9LRWd3s.png" alt="2D PCA with phenograph clusters" />
<p class="caption">2D PCA with phenograph clusters</p>
</div>
<p>The 2D PCA visualization doesn’t provide good visual separation of clusters. Adding an extra PC adds a significant amount of visual resolution to the clustering. Here is a toy example from a manually gated sample of PBMC from a healthy donor (toggle clusters by clicking on the names in the legend):</p>
<div class="figure">
<img src="3D--PCA--with--Phenograph--clusters" alt="{&quot;type&quot;:&quot;data&quot;, &quot;dataset&quot;:&quot;59afbf1da6b06402ace2194d&quot;, &quot;subset&quot;:&quot;null&quot;}" />
<p class="caption">{&quot;type&quot;:&quot;data&quot;, &quot;dataset&quot;:&quot;59afbf1da6b06402ace2194d&quot;, &quot;subset&quot;:&quot;null&quot;}</p>
</div>
<pre class="r"><code># Visualize results on t-SNE plot
getPalette = colorRampPalette(brewer.pal(9, \&quot;Set1\&quot;))
cols &lt;- getPalette(max(clusters_pg))
plot(tsne$Y[,1:2], pch=21, bg=alpha(col=cols[clusters_pg],0.6), col=\&quot;grey\&quot;, xlab=\&quot;\&quot;, ylab=\&quot;\&quot;, lwd=0.5)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/sbEQxYt.png" alt="2D t-SNE with phenograph clusters" />
<p class="caption">2D t-SNE with phenograph clusters</p>
</div>
<p>Below is an interactive implementation of Phenograph clustering + t-SNE on a sample of PBMC from a healthy donor (toggle clusters by clicking on the names in the legend):</p>
<div class="figure">
<img src="2D--t-SNE--with--Phenograph--clusters" alt="{&quot;type&quot;:&quot;data&quot;, &quot;dataset&quot;:&quot;59afc3b6a6b06402ace2194e&quot;, &quot;subset&quot;:&quot;null&quot;}" />
<p class="caption">{&quot;type&quot;:&quot;data&quot;, &quot;dataset&quot;:&quot;59afc3b6a6b06402ace2194e&quot;, &quot;subset&quot;:&quot;null&quot;}</p>
</div>
<p>If you want to see how multiple samples compare with clustering and t-SNE, one approach is to pool events from multiple samples, cluster them together, and plot them individually on a t-SNE canvas (recall that t-SNE will produce potentially vastly different layouts if samples are run separately). Below is an example with four samples:</p>
<pre class="r"><code># Sample 5000 random events from each sample
sub &lt;- NULL
for(i in 1:4) {
  temp &lt;- live[live$sample==fcs_files[i],]
  sub &lt;- rbind(sub, temp[sample(nrow(temp), 5000, replace=FALSE),])
}
# Check for duplicated events (remove if applicable)
table(duplicated(sub[,lineage_channels]))
sub &lt;- sub[-which(duplicated(sub[,lineage_channels]))]
# Cluster using phenograph
clusters_pg &lt;- cytof_cluster(xdata = sub[,lineage_channels], method = \&quot;Rphenograph\&quot;)
# Run t-SNE
set.seed(42)
tsne &lt;- Rtsne(sub[,lineage_channels])
# Plot each sample on t-SNE canvas consisting of all events
getPalette = colorRampPalette(brewer.pal(9, \&quot;Set1\&quot;))
cols &lt;- getPalette(max(clusters_pg))
par(mfrow=c(2,2))
for(i in 1:4) {
  plot(tsne$Y[,1:2], pch=19, col=\&quot;slategray4\&quot;, xlab=\&quot;\&quot;, ylab=\&quot;\&quot;, lwd=0.5, main=paste(\&quot;sample\&quot;, i))
  points(tsne$Y[sub$sample==unique(sub$sample)[i],1:2], pch=21, bg=alpha(col=cols[clusters_pg[sub$sample==unique(sub$sample)[i]]],0.6), col=\&quot;grey\&quot;, xlab=\&quot;\&quot;, ylab=\&quot;\&quot;, lwd=0.5)
}</code></pre>
<div class="figure">
<img src="https://i.imgur.com/u693Qgl.png" alt="Sample comparison using t-SNE" />
<p class="caption">Sample comparison using t-SNE</p>
</div>
</div>
<div id="flowsom-clustering" class="section level3">
<h3><span class="header-section-number">3.2.2</span> FlowSOM clustering</h3>
<p>FlowSOM is a method for clustering and visualization of flow and mass cytometry data. FlowSOM uses two levels of clustering to visualize how the intensity of cell type markers varies across the populations, and the amount of clusters used is quite high to ensure a visualization of even small subpopulations.</p>
<p>A standalone <a href="https://bioconductor.org/packages/release/bioc/html/FlowSOM.html">R package for FlowSOM</a> enables clustering, SOM, and MST visualizations. The package is built to run on FCS files rather than expression matrices, so here we will use the FlowSOM clustering implementation from the cytofkit package.</p>
<pre class="r"><code>library(cytofkit)
clusters_fs &lt;- cytof_cluster(xdata = live[live$sample==fcs_files[1],lineage_channels], method = \&quot;FlowSOM\&quot;, FlowSOM_k=25)</code></pre>
<pre class="r"><code># Visualize results on t-SNE plot
getPalette = colorRampPalette(brewer.pal(9, \&quot;Set1\&quot;))
cols &lt;- getPalette(max(clusters_fs))
plot(tsne$Y[,1:2], pch=21, bg=alpha(col=cols[clusters_fs],0.6), col=\&quot;grey\&quot;, xlab=\&quot;\&quot;, ylab=\&quot;\&quot;, lwd=0.5)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/QBZCQ7M.png" alt="2D t-SNE with FlowSOM cluster" />
<p class="caption">2D t-SNE with FlowSOM cluster</p>
</div>
</div>
<div id="vortex" class="section level3">
<h3><span class="header-section-number">3.2.3</span> VorteX</h3>
<p>VorteX is a clustering tool, which is implemented in Java and available as a graphical tool. VorteX includes different clustering algorithms incl. X-shift (non-parametric, density-based), hierarchical clustering, Mean-shift and K-medoids. In VorteX it is also easy to visualize the results or clusterings using different plot types.</p>
</div>
<div id="scaffold" class="section level3">
<h3><span class="header-section-number">3.2.4</span> SCAFFoLD</h3>
<p>SCAFFoLD (Single-Cell Analysis by Fixed Force- and Landmark-Directed) is, as the name indicates, a tool based on the use of force-directed graphs. First, cells are clustered, and then the clusters are spatialized in the plane. This means that similar clusters are put closer to each other than dissimilar ones. This results in a map. Furthermore, it is possible to highlight landmark populations that have been manually gated (working as references). As a result it is possible to compare a new sample to an old, to look for global structure changes. Samples can be clustered both individually and together.</p>
<p>To use SCAFFoLD, you need a C++ compiler and once that is ready you can run the program as follows:</p>
<pre class="r"><code># Loading required packages
require(devtools) || install.packages(\&quot;devtools\&quot;)
source(\&quot;http://bioconductor.org/biocLite.R\&quot;)
biocLite(\&quot;flowCore\&quot;)
library(devtools)
install_github(\&quot;nolanlab/scaffold\&quot;)
library(scaffold)

# Running SCAFFoLD (launching a GUI), please note that your current working directory should contain your files of interest before launching the GUI
scaffold.run()</code></pre>
<p>A complete use guide for the GUI is found <a href="https://github.com/nolanlab/scaffold">here</a>.</p>
</div>
</div>
<div id="visualizing-marker-expression-in-different-clusters" class="section level2">
<h2><span class="header-section-number">3.3</span> Visualizing marker expression in different clusters</h2>
<p>Once cells have been clustered, examining the expression profiles of the cells in each cluster can be used to provide clues about the type of cells constituting the clusters. This can be done in various ways:</p>
<div id="simple-heatmap" class="section level3">
<h3><span class="header-section-number">3.3.1</span> Simple heatmap</h3>
<p>The most basic way to visually show the marker profiles of clusters is with a simple heatmap of median marker expressions. This of course has the drawback of ignoring the variance of the marker expression in each cluster, but offers does offer some insights into the composition of each cluster.</p>
<pre class="r"><code>library(miscTools)
sub_matrix &lt;- live[live$sample==fcs_files[1],lineage_channels]
cluster_matrix &lt;- NULL
for(i in 1:max(clusters)) {
  cluster_matrix &lt;- rbind(cluster_matrix, colMedians(sub_matrix[clusters==i,]))
}
library(gplots)
cols = colorRampPalette(c(brewer.pal(9, \&quot;Set1\&quot;)[2], brewer.pal(9, \&quot;Set1\&quot;)[1]))
par(mar = c(2,2,2,2))
heatmap.2(t(cluster_matrix), col=cols, trace=\&quot;none\&quot;, density.info = \&quot;none\&quot;, sepcolor = \&quot;white\&quot;, sepwidth = c(0.001, 0.001), colsep=c(1:ncol(t(cluster_matrix))), rowsep=c(1:nrow(t(cluster_matrix))), xlab=\&quot;cluster\&quot;, ylab=\&quot;channel\&quot;, scale=\&quot;row\&quot;)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/fcAk8cO.png" alt="Simple heatmap of marker expression in each phenograph cluster" />
<p class="caption">Simple heatmap of marker expression in each phenograph cluster</p>
</div>
</div>
<div id="radviz" class="section level3">
<h3><span class="header-section-number">3.3.2</span> Radviz</h3>
<p>Radviz is a visualization scheme that utilizes so-called dimensional anchors. Each anchor (in this case, each marker) is distributed on the circumference of a circle, and ordered in a way that markers whose expression correlate, are located closer to each other. Each event is then projected into the circular and located closer to markers that best describe them. The authors use the following descriptive analogy to explain the algorithm: &quot;In Radviz, each dimension in the dataset is represented by a dimensional anchor, and each dimensional anchor is distributed evenly on a unit circle. Each line in the data set corresponds to a point in the projection, that is linked to every dimensional anchor by a spring. Each spring’s stiffness corresponds to the value for that particular thing in that particular dimension. The position of the point is defined as the point in the 2D space where the spring’s tension is minimum.&quot; <a href="https://cran.r-project.org/web/packages/Radviz/vignettes/single_cell_projections.html">Read more here</a>.</p>
<pre class="r"><code>library(Radviz)
cell.S &lt;- make.S(colnames(sub_matrix))
cell.sim &lt;- cosine(as.matrix(sub_matrix))
optim.cell &lt;- do.optim(cell.S,cell.sim,iter=100,n=1000)
cell.S &lt;- make.S(tail(optim.cell$best,1)[[1]])
cell.rv &lt;- do.radviz(as.matrix(sub_matrix),cell.S)
layout(matrix(c(1,1,1,2), nrow=1,ncol=4))
plot(cell.rv, point.shape=19, point.color=alpha(cols[clusters], 0.6))
plot.new()
legend(\&quot;center\&quot;, legend=paste(\&quot;cluster\&quot;, 1:max(clusters)), col=cols, cex=1, pch=16, bty=&#39;n&#39;)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/Sm84OCA.png" alt="Radviz plot" />
<p class="caption">Radviz plot</p>
</div>
<p>Clusters can be visualized more clearly in a &quot;bubble plot&quot; version of Radviz. In this plot, events are collapsed to the coordinates of their respective cluster centroid, where the size of the &quot;bubble&quot; corresponds to the size of the cluster.</p>
<pre class="r"><code>pop.rv &lt;- do.radviz(cluster_matrix,cell.S)
size &lt;- table(clusters)
layout(matrix(c(1,1,1,2), nrow=1,ncol=4))
bubbleRadviz(pop.rv, bubble.color=alpha(cols[1:nrow(cluster_matrix)],0.8), bubble.size=log(size[1:nrow(cluster_matrix)]), scale=0.2, decreasing=TRUE)
plot.new()
legend(\&quot;center\&quot;, legend=paste(\&quot;cluster\&quot;, 1:max(clusters)), col=cols, cex=1, pch=16, bty=&#39;n&#39;)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/ry5Fffs.png" alt="Radviz bubble plot" />
<p class="caption">Radviz bubble plot</p>
</div>
</div>
</div>
<div id="analysis-of-differential-abundance-of-cell-populations-and-proteins-daanalysis" class="section level2">
<h2><span class="header-section-number">3.4</span> Analysis of differential abundance of cell populations and proteins {daanalysis}</h2>
<p>Once subpopulations are clustered in each sample, the next step will often be to elucidate associations between cluster frequencies and/or protein expression and a given condition or time. These exercises closely resemble differential expression analyses used in transcriptomics for decades, and the bioinformatics toolbox contains multiple tools for basic significance tests and linear models. In addition, the GUI tools Citrus and Statistical Scaffold were developed for this purpose.</p>
<p>Citrus (cluster identification, characterization, and regression) (Bruggner et al. 2014) works by combining hierarchical clustering of events in all samples aggregated into one. Then, descriptive metrics for each cluster are calculated, including percentage of cells assigned to each cluster, and the median expression of lineage markers in each cluster. These data are then used in regularized classification models to elucidate factors differing across conditions.</p>
<p>The Statistical Scaffold method (Spitzer et al. 2017) works in a similar way. First, populations of interest are separated by manual gating in a representative sample. These populations will serve as landmark nodes on a scaffold map, which serves as a template for mapping events in the remaining samples using the cosine similarity between the vectors of median marker values of each cluster. Lastly, the Significance Across Microarrays permutation test (Tusher et al. 2001) is applied to test for differences in population frequencies across conditions.</p>
<p>One major shortcoming of existing tools is the inability to handle more complex experimental setups, such as longitudinal studies, or studies with multiple factors and covariates. Generalized linear models (GLM) for differential expression analysis, popularized for transcriptomic analyses by tools such as limma (Ritchie et al. 2015) for microarray data and DEseq2 (Love et al. 2014) or EdgeR (Robinson et al. 2010) for RNA sequencing data, accepts model matrices with co-variates for complex experimental designs. Recently, Lun and colleagues leveraged GLMs from the EdgeR software specifically for mass cytometry data and showed these models outperform Citrus on legacy data (Lun et al. 2017).</p>
<p>Lastly, it is important to consider that the immune system consists of multiple functional cell populations, which are coordinated in a complex interplay in response to various conditions. It is therefore sometimes more informative to examine population frequencies as dependent variables by examining correlations or ratios between frequencies (Bailur et al. 2017). Additionally, natural variations in the immune system are likely to mask the difference conferred by time or conditions when looking at a single variable, even if analyzing a large number of samples. Much like transcriptomics analyses, the immune system does not consist of independent variables, and neither the abundance of protein markers on individual cells nor the frequencies of individual cell populations can be assumed independent from the remaining variables.</p>
<div id="analysis-of-differential-abundance-of-cells-in-clusters-dacells" class="section level3">
<h3><span class="header-section-number">3.4.1</span> Analysis of differential abundance of cells in clusters {dacells}</h3>
<p>The cell counts in each cluster is the easiest to compare, as this roughly follows a negative binomial distribution and can be flexibly modeled using a generalized linear model (GLM) as implemented in, for example, the edgeR package for RNA-seq transcriptomics data. In this example, we will use the data generated by Bodenmiller et al. and identify clusters with a differential abundance of cells between conditions.</p>
<p>First, we import the data and create a dataframe with all the samples:</p>
<pre class="r"><code># Import data
library(flowCore)
setwd(\&quot;~/cytof_primer/case_control/fcs_files/\&quot;)
fcs_files &lt;- list.files(pattern=&#39;.fcs$&#39;, full=TRUE, ignore.case = TRUE)
exprs_set &lt;- data.frame()
sample &lt;- c()
for(i in fcs_files) {
  fcs &lt;- read.FCS(filename=i, transformation=FALSE) 
  exprs &lt;- fcs@exprs
  colnames(exprs) &lt;- gsub(pattern = \&quot;.*_\&quot;, replacement = \&quot;\&quot;, x = as.vector(fcs@parameters@data$desc))
  exprs &lt;- asinh(exprs/5)
  exprs_set &lt;- rbind(exprs_set, exprs)
  sample &lt;- append(sample, rep(i, nrow(exprs)))
}
exprs_set$sample &lt;- sample

# Set channels
pregating_channels &lt;- c(\&quot;DNA-1\&quot;, \&quot;DNA-2\&quot;)
lineage_channels &lt;- c(\&quot;CD45\&quot;, \&quot;CD4\&quot;, \&quot;CD20\&quot;, \&quot;CD33\&quot;, \&quot;CD123\&quot;, \&quot;CD14\&quot;, \&quot;IgM\&quot;, \&quot;HLA-DR\&quot;, \&quot;CD7\&quot;, \&quot;CD3\&quot;)
functional_channels &lt;- c(\&quot;pNFkB\&quot;, \&quot;pp38\&quot;, \&quot;pStat5\&quot;, \&quot;pAkt\&quot;, \&quot;pStat1\&quot;, \&quot;pSHP2\&quot;, \&quot;pZap70\&quot;, \&quot;pStat3\&quot;, \&quot;pSlp76\&quot;, \&quot;pBtk\&quot;, \&quot;pPlcg2\&quot;, \&quot;pErk\&quot;, \&quot;pLat\&quot;, \&quot;pS6\&quot;)
instrument_channels &lt;- c(\&quot;length\&quot;, \&quot;Time\&quot;, \&quot;sample\&quot;, \&quot;BC1\&quot;, \&quot;BC2\&quot;, \&quot;BC3\&quot;, \&quot;BC4\&quot;, \&quot;BC5\&quot;, \&quot;BC6\&quot;, \&quot;BC7\&quot;)</code></pre>
<p>In the interest of speed, we here down sample to 1000 random events from each sample. Remember to normalize to the total number of events in each sample if the full samples are used. We then cluster using the cytofkit implementation of phenograph and create a table with cell counts in each cluster for each sample.</p>
<pre class="r"><code># Sample 1000 random events form each sample
sub &lt;- NULL
for(i in unique(exprs_set$sample)) {
  temp &lt;- exprs_set[exprs_set$sample==i,]
  sub &lt;- rbind(sub, temp[sample(nrow(temp), 1000, replace=FALSE),lineage_channels])
}
sample &lt;- rep(unique(exprs_set$sample), each=1000)

# cluster
library(cytofkit)
clusters_pg &lt;- cytof_cluster(xdata = sub, method = \&quot;Rphenograph\&quot;)

# make frequency table
cell_counts &lt;- matrix(nrow=max(clusters_pg), ncol=0)
for (i in unique(sample)) {
  sample_count &lt;- table(clusters_pg[sample==i])
  column &lt;- unname(sample_count[match(1:nrow(cell_counts), names(sample_count))])
  column[is.na(names(sample_count[match(1:nrow(cell_counts), names(sample_count))]))] &lt;- 0
  cell_counts &lt;- cbind(cell_counts, column)
}
conditions &lt;- rep(1, 16)
conditions[grep(\&quot;BCR-XL\&quot;, fcs_files)] &lt;- \&quot;BCR-XL\&quot;
conditions[grep(\&quot;Reference\&quot;, fcs_files)] &lt;- \&quot;Reference\&quot;
colnames(cell_counts) &lt;- conditions</code></pre>
<p>Lastly, we evaluate differential abundance in each cluster using a NB-GLM from the edgeR package:</p>
<pre class="r"><code># calculate differential abundance
library(edgeR)

dge &lt;- DGEList(cell_counts, lib.size=rep(1000,16))
design &lt;- model.matrix(~factor(conditions))
y &lt;- estimateDisp(dge, design)
fit &lt;- glmQLFit(y, design, robust=TRUE)
res &lt;- glmQLFTest(fit, coef=2)</code></pre>
<p>We can now visualize the log2 fold change in clusters with a significant differential abundance between the two conditions:</p>
<pre class="r"><code># visualize on t-SNE
library(Rtsne)
library(RColorBrewer)
library(scales)

set.seed(42)
tsne &lt;- Rtsne(sub)

# par(mfrow=c(1,2))
layout(matrix(c(1,1,1,2,2,2,3), nrow=1,ncol=7))

getPalette = colorRampPalette(brewer.pal(9, \&quot;Set1\&quot;))
cols &lt;- getPalette(max(clusters_pg))
plot(tsne$Y, pch=21, bg=alpha(col=cols[clusters_pg],0.6), col=\&quot;grey\&quot;, xlab=\&quot;\&quot;, ylab=\&quot;\&quot;, lwd=0.5, main=\&quot;Phenograph clusters\&quot;)

log2fc &lt;- res$table$logFC
log2fc[res$table$PValue&gt;0.05] &lt;- 0
colpal &lt;- colorRampPalette(c(&#39;red&#39;, \&quot;grey\&quot;, &#39;blue&#39;))
cols &lt;- colpal(100)[as.numeric(cut(log2fc,breaks = 100))]
plot(tsne$Y, pch=21, bg=alpha(col=cols[clusters_pg],0.6), col=\&quot;grey\&quot;, xlab=\&quot;\&quot;, ylab=\&quot;\&quot;, lwd=0.5, main=\&quot;colored by log2FC\&quot;)

legend_image &lt;- as.raster(matrix(colpal(20), ncol=1))
plot(c(0,2),c(0,1),type = &#39;n&#39;, axes = F,xlab = &#39;&#39;, ylab = &#39;&#39;, main = &#39;log2FC&#39;)
text(x=1.5, y = seq(0,1,l=3), labels = c(round(min(log2fc), digits = 2),0,round(max(log2fc),digits=2)))
rasterImage(legend_image, 0, 0, 1, 1)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/JRscapE.png" alt="Visualizing differential abundance on t-SNE plots" />
<p class="caption">Visualizing differential abundance on t-SNE plots</p>
</div>
<p>We can also visually show the cell abundance difference between clusters using the approach in <a href="#clustering">section 3.2</a>:</p>
<div class="figure">
<img src="https://i.imgur.com/zwkBhok.png" alt="Visualizing cell abundance on t-SNE canvas" />
<p class="caption">Visualizing cell abundance on t-SNE canvas</p>
</div>
<div id="cydar" class="section level4">
<h4><span class="header-section-number">3.4.1.1</span> cydar</h4>
<p>The cydar tool works in much the same way as above - clustering followed by statistical testing for differential abundance of cells in each cluster. In this method, the clustering is done by creating hyperspherical clusters. This partitioning is similar to that of k-means clustering, but here a constant distance measure (the radius) is used to define the hyperspheres rather than dividing the events into Voronoi cells based on the nearest mean. This means that rather than getting a (somewhat) interpretable cluster structure (where each cluster can be labeled as a canonical cell population), hyperspherical partitioning is somewhat abstract and not directly interpretable. Another feature of the hypersphere partitioning is allowing multi-membership (which one can argue makes more biological sense than strictly allowing cells to belong to one cluster), but this again makes the partitioning harder to interpret in 2 or 3 dimensional space. Although harder to directly interpret, choosing a k that is greater than &quot;reality&quot;, is certainly better than selecting a k that is too small: &quot;under-clustering&quot; a highly heterogeneous cell mixture such as PBMC, will potentially result in comparing biologically irrelevant subsets. However, the trade-off of erring on the safe side is that in order for the cydar output to be biologically interpretable, some post-processing is necessary. The authors provide an excellent discussion of this in their <a href="https://www.nature.com/nmeth/journal/v14/n7/full/nmeth.4295.html">paper</a> and in their <a href="https://bioconductor.org/packages/release/bioc/vignettes/cydar/inst/doc/cydar.html">R package vignette</a></p>
<p>Now for some code!</p>
<pre class="r"><code>library(cydar)
library(ncdfFlow)

# Sample 1000 random events form each sample - cydar takes data as a list of matrices
sub &lt;- list()
for(i in unique(exprs_set$sample)) {
  temp &lt;- exprs_set[exprs_set$sample==i,]
  sub[[i]] &lt;- temp[sample(nrow(temp), 1000, replace=FALSE),lineage_channels]
}
conditions &lt;- rep(1, 16)
conditions[grep(\&quot;BCR-XL\&quot;, fcs_files)] &lt;- \&quot;BCR-XL\&quot;
conditions[grep(\&quot;Reference\&quot;, fcs_files)] &lt;- \&quot;Reference\&quot;
sample.id &lt;- rep(c(1:16), each=1000)

# Count the cells into hyperspheres with a radius equal to 0.5 times the square root of the number of markers
cd_all &lt;- prepareCellData(sub)
cd_all &lt;- countCells(cd_all, tol=0.5)

# Testing for significant differences in abundance
library(edgeR)
y &lt;- DGEList(assay(cd_all), lib.size=cd_all$totals)
keep &lt;- aveLogCPM(y) &gt;= aveLogCPM(5, mean(cd_all$totals))
cd &lt;- cd_all[keep,]
y &lt;- y[keep,]
design &lt;- model.matrix(~factor(conditions))
y &lt;- estimateDisp(y, design)
fit &lt;- glmQLFit(y, design, robust=TRUE)
res &lt;- glmQLFTest(fit, coef=2)

# Controlling the spatial FDR
qvals &lt;- spatialFDR(intensities(cd), res$table$PValue)
is.sig &lt;- qvals &lt;= 0.05</code></pre>
<p>In this case we find that 253 out of 549 hyperspheres contain a significantly different number of cells between our two conditions.</p>
<pre class="r"><code># Print the number of hypersphere with significantly different cell abundance between case and control
summary(is.sig)
   Mode   FALSE    TRUE 
logical     296     253 </code></pre>
</div>
</div>
<div id="analysis-of-differential-expression-of-proteins-between-clusters-deprot" class="section level3">
<h3><span class="header-section-number">3.4.2</span> Analysis of differential expression of proteins between clusters {deprot}</h3>
<p>The expression of lineage markers between different clusters will naturally differ, as differential expression is the very basis for all the clustering algorithms discussed here. However, it may be of interest to examine whether the expression of functional markers differs within a cluster that is considered to represent the same cell type in two different conditions.</p>
<p>Currently, the most widely utilized approach is to compare the median expression a given protein in each sample across conditions for a specific cluster. There are some potentially <strong>serious caveats to this approach</strong> which we will discuss a bit further down. But first, let’s go through a simple example of this common approach:</p>
<pre class="r"><code># Sample 1000 random events form each sample
sub &lt;- NULL
for(i in unique(exprs_set$sample)) {
  temp &lt;- exprs_set[exprs_set$sample==i,]
  sub &lt;- rbind(sub, temp[sample(nrow(temp), 1000, replace=FALSE),c(lineage_channels, functional_channels)])
}
sample &lt;- rep(unique(exprs_set$sample), each=1000)

# cluster
library(cytofkit)
clusters_pg &lt;- cytof_cluster(xdata = sub[,lineage_channels], method = \&quot;Rphenograph\&quot;)

# Make median expression table (samples x functional proteins) for each cluster
# - here we generate the table for cluster 1
library(miscTools)
sub_cluster &lt;- sub[clusters_pg==1,functional_channels]
sample_cluster &lt;- sample[clusters_pg==1]
exprs_cluster &lt;- NULL
for(i in unique(sample_cluster)) {
  exprs_cluster &lt;- cbind(exprs_cluster, colMedians(sub_cluster[sample_cluster==i,]))
}
colnames(exprs_cluster) &lt;- unique(sample_cluster)</code></pre>
<p>Again we are left with a simple expression matrix, to which you can apply your favorite statistics. Here we use a linear model:</p>
<pre class="r"><code># Basic DE test using limma
library(limma)
conditions_test &lt;- factor(gsub(\&quot;.*(Reference|BCR-XL).*\&quot;, \&quot;\\\\1\&quot;, colnames(exprs_cluster)))
patient_test &lt;- factor(gsub(\&quot;.*patient(.)_.*\&quot;, \&quot;\\\\1\&quot;, colnames(exprs_cluster)))
design &lt;- model.matrix(~patient_test+conditions_test)
fit &lt;- lmFit(exprs_cluster, design)
fit &lt;- eBayes(fit)
res &lt;- topTable(fit, coef=\&quot;conditions_testReference\&quot;, number=\&quot;inf\&quot;, adjust.method=\&quot;bonferroni\&quot;)</code></pre>
<p>Which results in:</p>
<pre><code>              logFC     AveExpr           t      P.Value    adj.P.Val         B
pNFkB   0.901599304  1.64626641  8.95197298 1.570931e-05 0.0002199304  3.567493
pAkt   -0.495415877  2.57505470 -4.62954456 1.553046e-03 0.0217426438 -1.285979
pSlp76  0.277175958  0.07675553  3.97151164 3.855769e-03 0.0539807621 -2.245577
pBtk    0.416567743  0.16310906  3.73853120 5.400805e-03 0.0756112746 -2.599032
pp38    0.492161765  0.17251180  3.27901472 1.072462e-02 0.1501446325 -3.313240
pErk   -0.245753542  0.32830765 -3.08944667 1.433815e-02 0.2007340740 -3.612798
pLat   -0.226744720  0.28252997 -2.72364712 2.534997e-02 0.3548995620 -4.194273
pS6    -0.187262388  0.06369062 -2.70666148 2.603589e-02 0.3645024835 -4.221272
pStat3 -0.152212475  0.08025186 -2.70645026 2.604454e-02 0.3646235574 -4.221607
pStat1 -0.341139355  1.00338611 -2.18346123 5.949445e-02 0.8329222973 -5.042568
pZap70 -0.066471291 -0.02515855 -2.13249427 6.447399e-02 0.9026358745 -5.120647
pPlcg2 -0.211360381  0.55087960 -2.07229530 7.088211e-02 0.9923495922 -5.212186
pStat5 -0.009597480 -0.06214086 -0.49439836 6.339071e-01 1.0000000000 -7.001390
pSHP2  -0.001402016 -0.06092106 -0.07812268 9.395902e-01 1.0000000000 -7.132165</code></pre>
<p>So in cluster 1 we have two significantly differentially expressed proteins between the Reference and BCR-XL samples, based on the median expression of the protein.</p>
<p>The most glaring (potential) issue with using the median expression is that we assume homoscedasticity and ignore the variance of the expression within each sample. Another potential issue is that for the linear models, we assume that the expression values (and median values) follow a Gaussian distribution. As mentioned earlier, the arcsinh transformation is first and foremost a cosmetic operation for biaxial plots, and the raw counts (without the default randomization by the Fluidigm software) roughly follows a (somewhat zero-inflated) negative binomial distribution. As such, a more appropriate statistic may be adopted from common practices in single cell transcriptomics analyses, but the error model has to be adapted: for example, dropout events are not an issue in mass cytometry, but the isotope counts are subject to stochastic noise and antibody binding introduces some fairly systematic noise (as discussed in our manuscript) that should be modeled appropriately. Lastly, we measure far fewer features in mass cytometry, but vastly more observations compared with single cell transcriptomics. Until such a model is formulated, linear modeling of median marker expression remains the state-of-the-art.</p>
<div id="citrus" class="section level4">
<h4><span class="header-section-number">3.4.2.1</span> Citrus</h4>
<p>One of the tools that utilizes this method is Citrus (cluster identification, characterization, and regression). The Citrus algorithm works by first doing an optional down-sampling to a user-specified number of events from each sample followed by hierarchical clustering of the pooled cells. When clusters are defined, Citrus utilizes algorithms from the SAM (statistical analysis of microarrays) package to calculate 1) differential abundance of cells in clusters across the defined conditions, 2) differential median protein expression in each cluster across the defined conditions, and 3) trains a lasso regularized logistic regression and nearest shrunken centroids method to identify a subset of cluster features that best stratifies the condition from the control samples. Citrus is primarily implemented as a GUI, but an <a href="https://github.com/nolanlab/citrus">R package</a> allow for command-based execution. In either case, Citrus takes only FCS files. The method is described in the <a href="http://www.pnas.org/content/111/26/E2770.long">2014 paper by Bruggner et al.</a>.</p>
</div>
</div>
</div>
<div id="cell-type-assignment-celltypes" class="section level2">
<h2><span class="header-section-number">3.5</span> Cell type assignment {celltypes}</h2>
<p>In order to finally draw biological conclusions from the clustering and differential population frequency and/or protein expression analyses, differential population frequency analyses, or differential marker expression analyses, it is necessary to assign a cell type to each cluster. For domain experts, this may be as simple as comparatively surveying the marker expression in each cluster, for example by generating clusters vs. markers heatmaps or aided by tools such as Marker Enrichment Modelling (MEM) (Diggins et al. 2017), which can be used to label clusters with a simpler human readable marker enrichment profile. Other tools such as flowCL (Courtot et al. 2015), ACDC (Lee et al. 2017) or the LDA-approach by <a href="https://doi.org/10.1002/cyto.a.23738">Abdelaal et al. (2019)</a> enables automated assignment of labels to clusters, but flowCL requires the user to make decisions about when a marker is expressed or not and uses the phenotype label to assign the most likely label, and ACDC requires users to define population labels with a markers x populations binary expression matrix.</p>
<p>The lack of a fully automated cell type assignment workflow is most likely due to the fact that assigning canonical populations is trivial for domain experts, while rare populations may be subject to disagreement, and novel populations not yet named are likely to be present. However, for the computationally inclined, the Python implementation of ACDC can be used to perform a label estimate on clustered data.</p>
<div id="acdc" class="section level3">
<h3><span class="header-section-number">3.5.1</span> ACDC</h3>
<p>Automated Cell-type Discovery and Classiffcation (ACDC) takes a cell type-marker table, in which each cell type is related to a marker profile. The table can contain values 1 (present), -1 (absent), 0/NA (do not consider) for each marker and cell type combination. This table can be completely customized to suit the measured markers of a study and include the cell types one wishes to study. The values should be based on the biological knowledge of marker expression of different cell types (e.g. all T cells express CD3). Then each cell is classified resulting in a clustering in which each cluster is labeled as a specific cell type. This method is consequently biased by the cell-type marker table applied, but it may be useful in certain situations. ACDC is available for Python and tutorials for running the program can be found <a href="https://github.com/howchihlee/ACDC_tutorial/tree/master/notebooks">here</a>. We have also made <a href="https://github.com/biosurf/CyTOF/blob/master/ACDC_BioSurf.ipynb">our own IPython Notebook-tutorial</a> for running ACDC.</p>
<p>Three different cell-type marker tables can be accessed in our Git repository: The <a href="https://github.com/biosurf/CyTOF/blob/master/AML_table.csv">AML table</a>, the <a href="https://github.com/biosurf/CyTOF/blob/master/BMMC_table.csv">BMMC table</a>, and a <a href="https://github.com/biosurf/CyTOF/blob/master/Leipold_table.csv">table defined by Michael D. Leipold</a>. The two first tables were derived from <a href="https://www.cytobank.org/nolanlab/reports/Levine2015.html">Levine et al.</a> and processed into the csv format by the original ACDC-developers.</p>
</div>
<div id="marker-enrichment-modeling" class="section level3">
<h3><span class="header-section-number">3.5.2</span> Marker Enrichment Modeling</h3>
<p>The Marker Enrichment Modeling (MEM) tools describes cellular subsets by the features that are enriched for each population using a weighted marker expression profile. MEM describes the characteristics of different clusters in the context of the other cells in the sample - e.g. certain surface markers specifically enriched in the given cluster.</p>
<p>The output is a weighted marker expression profile of each cluster, which can be translated into a &quot;human readable&quot; label for each cluster. In this example, we use the clusters from phenograph and create a list of human readable labels for each cluster.</p>
<pre class="r"><code># Calculate enrichment scores
mem &lt;- MEM(cbind(live[live$sample==unique(live$sample)[1],lineage_channels], cluster&lt;-clusters_pg))

# This creates a table of enrichment scores for each marker in each cluster. Round these scores to get a label as presented in their paper
mem &lt;- round(mem[[5]][[1]])

# Adapt label format from numeric to \&quot;--\&quot;, \&quot;-\&quot;, \&quot;+\&quot;. \&quot;++\&quot; (this is necessary for the \&quot;flowcl\&quot; package below)
mem &lt;- mem[order(as.numeric(rownames(mem))),]
threshold &lt;- median(abs(as.vector(mem)))
labels &lt;- list()
for(i in 1:nrow(mem)) {
  lab &lt;- mem[i,abs(mem[i,])&gt;0]
  lab[lab &lt; -threshold] &lt;- \&quot;--\&quot;
  lab[lab &lt; 0] &lt;- \&quot;-\&quot;
  lab[lab &gt; threshold] &lt;- \&quot;++\&quot;
  lab[lab &gt; 0] &lt;- \&quot;+\&quot;
  lab &lt;- paste(names(lab), lab, sep=\&quot;\&quot;)
  labels[[rownames(mem)[i]]] &lt;- paste(lab, collapse=\&quot;\&quot;)
}</code></pre>
<p>This creates the following cluster labels:</p>
<pre><code>[1] CD19+CD4++CD11c++CD16+CD3++CD38++CD27++CD14+CXCR5+CCR7+CD20+CD127++CD33++CD28++CD123+HLADR++CD25+
[2] CD19-CD4++CD8-CD11c++CD16+CD3-CD38++CD27-CXCR5+CCR7-CD45RA-CD127-CD33++CD28++CD123+HLADR++CD25++
[3] CD4++CD8-IgD-CD11c-CD16-CD3++CD38-CD27++CD14-CXCR5++CCR7++CD45RA-CD20-CD127++CD33+CD28++CD161+CD123-CD56-HLADR-
[4] CD19-CD4-IgD-CD11c-CD16+CD3-CD38+CD27-CD14-CXCR5-CCR7-CD45RA+CD20-CD127-CD33-CD28-CD161+CD123+CD56+HLADR-CD25-
[5] CD4-CD11c-CD16-CD3++CD38-CD27++CD14-CXCR5-CCR7-CD45RA-CD20-CD127++CD33+CD28++CD161+CD123+HLADR-CD25-
[6] CD57+CD19-CD4++CD8-CD11c++CD16+CD3-CD38++CD27-CD14++CXCR5+CCR7-CD45RA-CD127-CD33++CD28-CD161+CD56+HLADR++CD25+
[7] CD4++CD8-CD11c-CD16-CD3++CD38-CD27++CD14-CXCR5+CCR7++CD45RA+CD20-CD127+CD33+CD28++CD123+CD56-HLADR-CD25+
[8] CD57+CD4++CD8-CD11c-CD16-CD3++CD38+CD27++CD14+CXCR5-CCR7++CD45RA+CD20-CD127++CD33+CD28++CD161+CD123-CD56-HLADR-CD25-
[9] CD4++CD8-IgD-CD11c-CD16-CD3++CD38-CD27-CD14-CXCR5-CCR7-CD45RA-CD20-CD127++CD33+CD28++CD123+CD56-HLADR-CD25+
[10] CD19-CD4-CD8-IgD-CD11c-CD16-CD3++CD27++CD14-CXCR5+CCR7+CD45RA++CD20+CD127+CD33-CD28++CD123+CD56-HLADR-CD25-
[11] CD4++CD8-IgD-CD11c-CD16-CD3++CD38-CD27++CD14-CXCR5-CCR7+CD45RA-CD20-CD127++CD33+CD28++CD123-CD56-HLADR-
[12] CD19++CD4+CD8+IgD++CD11c++CD16-CD3++CD38+CD27++CD14-CXCR5++CCR7++CD45RA++CD20++CD127++CD33-CD28++CD56+HLADR+CD25++
[13] CD19-CD4-CD8-IgD-CD11c++CD16-CD3-CD38-CD27-CXCR5-CCR7-CD45RA-CD20-CD127-CD33+CD28-CD123+CD56-HLADR++CD25+
[14] CD19++CD4-CD8-IgD++CD11c-CD16-CD3-CD38+CD27-CD14-CXCR5++CCR7+CD45RA++CD20++CD127-CD33-CD28-CD161+CD56+HLADR++CD25+
[15] CD4++CD8-IgD-CD11c-CD16-CD3++CD38-CD27++CD14-CXCR5-CCR7+CD45RA-CD20-CD127++CD33+CD28++CD161+CD123+CD56-HLADR-CD25+
[16] CD19-CD4++CD8-IgD-CD11c-CD16-CD3-CD38+CD27-CD14-CXCR5-CCR7+CD45RA++CD20+CD127-CD33-CD28-CD123++CD56+HLADR++CD25+
[17] CD57+CD4+CD8+CD11c+CD16++CD3++CD38+CD27++CCR7+CD45RA++CD127+CD33+CD28++CD161+CD123+CD56++HLADR-
[18] CD19++CD4-CD8-CD11c-CD16-CD3-CD38-CD27++CD14-CXCR5++CCR7+CD45RA++CD20++CD127-CD33-CD28-CD123+CD56-HLADR++CD25+
[19] CD4++CD8-IgD-CD11c-CD16-CD3++CD38-CD27+CD14-CXCR5-CCR7-CD45RA-CD20-CD127+CD33+CD28++CD123+CD56-HLADR++CD25+
[20] CD19-CD4-CD8-IgD-CD11c+CD16++CD3-CD38+CD27+CD14-CXCR5-CCR7-CD45RA++CD127-CD33-CD28-CD161+CD123+CD56++HLADR-CD25-
[21] CD4-CD8++CD11c-CD16-CD3++CD38-CD27++CXCR5-CCR7++CD45RA++CD20+CD127++CD33-CD28++CD123-CD56-HLADR-CD25-
[22] CD57++CD19-CD4-CD8+IgD-CD11c+CD16++CD3-CD38+CD27-CD14-CXCR5-CCR7-CD45RA++CD127-CD33-CD28-CD161+CD123-CD56++HLADR-CD25-
[23] CD4++CD8+CD11c++CD16++CD3-CD38-CD27-CXCR5+CCR7-CD45RA+CD20+CD127-CD33+CD28-CD161+CD123+CD56+HLADR++CD25+
[24] CD57++CD4-CD8++CD11c-CD3++CD38-CD27-CD14-CXCR5-CCR7+CD45RA-CD20-CD127-CD33-CD28-CD123-CD56-HLADR-CD25-
[25] CD4-CD8++CD11c-CD16-CD3++CD38-CD27++CD14-CXCR5-CCR7+CD45RA-CD20-CD127+CD33-CD28++CD123-CD56-HLADR-CD25-
[26] CD19-CD4-CD8+IgD-CD11c+CD16++CD3-CD38+CD27-CD14-CXCR5-CCR7-CD45RA++CD127-CD33-CD28-CD161+CD123+CD56++HLADR-CD25-</code></pre>
<p>These labels alone will not be meaningful to non-domain experts, and in order to assign a cell name to each cluster, one will need to query an ontology database.</p>
</div>
<div id="flowcl" class="section level3">
<h3><span class="header-section-number">3.5.3</span> flowCL</h3>
<p>flowCL assigns a phenotype to a cell based on its expressed markers - it matches against the <a href="http://cellontology.org">Cell Ontology (CL)</a>. An example query could be &quot;CD16-CD14+CD33+&quot;, but it is also possible to use – and ++ for the marker levels to be more specific regarding the level of a marker.</p>
<p>We can run flowCL directly on the list created using MEM:</p>
<pre class="r"><code>library(\&quot;flowCL\&quot;)
res &lt;- flowCL(MarkerList = labels, ExpMrkrLst = as.list(lineage_channels), VisualSkip = TRUE)</code></pre>
<p><strong>NOTE: flowCL is currently times out when pinging the Cell Ontology database. The authors have been contacted and are working on fixing the problem</strong></p>
</div>
</div>
<div id="cellular-hierarchies-hierarchies" class="section level2">
<h2><span class="header-section-number">3.6</span> Cellular hierarchies {hierarchies}</h2>
<p>Cells of the immune system exist in a hierarchy of different developmental and activation states. Estimating and visualizing the proportion of cells in a complex mixture belonging to different states can be useful to identify aberrations in cellular differentiation between different conditions. The spanning-tree progression analysis of density-normalized events (SPADE) (Qiu et al. 2011) is one of the first tools to address this problem for mass cytometry data. At the core of SPADE is a minimum spanning-tree algorithm, which is applied to clusters of cells to derive possible hierarchies within the data. Events are clustered using hierarchical clustering after data is downsampled in a density-dependent manner. This serves to speed up the clustering, but also ensures equal representation of rare and dense populations. The clusters in the resulting minimum spanning-tree must then be manually labelled, but this is aided by plots highlighted by marker intensity for each sample. Early versions of SPADE relied on stochastic downsampling and stochastic minimum spanning-trees, which produced different run-to-run results, but recent updates of the software produces deterministic results (Qiu 2017). SPADE is implemented in C++, but exists with an R wrapper.</p>
<p>Wanderlust is a graph-based approach to uncovering cell development trajectories. It works by first constructing a k-nearest neighbor graph, connecting each cell to its k nearest neighbors based on Euclidean distance. Then, the shortest path between an initiator node (a user selected root cell) and every other cell in the network is used to estimate the development trajectory of each event (Bendall et al. 2014). Wanderlust is implemented in Matlab.</p>
<div id="flowsom" class="section level3">
<h3><span class="header-section-number">3.6.1</span> FlowSOM</h3>
<p>Self-organizing maps (SOM) and minimum spanning tree (MST) visualizations can be generated using the kohonen package for the SOM, and the igraph package for the MST. The flowSOM implementations of these do contain some extra plotting options. If you want visualizations exactly like those presented in the <a href="http://onlinelibrary.wiley.com/doi/10.1002/cyto.a.22625/full">paper</a>, you may try their <a href="https://bioconductor.org/packages/release/bioc/html/FlowSOM.html">R package</a>. Note that it only takes FCS files as input, but individual functions can be edited if an explicit need for these visualizations exists.</p>
<pre class="r"><code># Calculate and visualize self organizing map with the parameters used by FlowSOM
library(kohonen)
som &lt;- som(as.matrix(live[live$sample==fcs_files[1],lineage_channels]), grid=somgrid(xdim = 10, ydim = 10), dist.fcts=\&quot;euclidean\&quot;)

# Calculate codebook clusters
library(ConsensusClusterPlus)
nclust &lt;- 15
cclust &lt;- ConsensusClusterPlus(t(som$codes[[1]]), maxK = nclust, reps = 100, pItem = 0.9, pFeature = 1, title = tempdir(), plot = \&quot;pdf\&quot;, verbose = FALSE, clusterAlg = \&quot;hc\&quot;, distance = \&quot;euclidean\&quot;, seed = NULL)
codebook_clusters &lt;- cclust[[nclust]]$consensusClass
codebook_list &lt;- list()
for(i in 1:max(codebook_clusters)) {
  codebook_list[[i]] &lt;- names(codebook_clusters[codebook_clusters==i])
}

# Create minimum spanning tree from codebook distance matrix
library(igraph)
graph  &lt;- graph.adjacency(as.matrix(dist(som$codes[[1]])), weighted = TRUE)
mst &lt;- mst(graph)

# Plot flowSOM-like MST output
layout &lt;- layout.kamada.kawai(as.undirected(mst))
marker_palette = colorRampPalette(brewer.pal(9, \&quot;Set1\&quot;))
cluster_palette = colorRampPalette(brewer.pal(8, \&quot;Set2\&quot;))
marker_cols &lt;- marker_palette(length(lineage_channels))
cluster_cols &lt;- cluster_palette(max(codebook_clusters))
medians &lt;- t(apply(som$codes[[1]], 1, function(x) x/sum(x)))
plot(as.undirected(mst), vertex.shape=\&quot;star\&quot;, vertex.label = NA, vertex.size = rep(10, length(codebook_clusters)), vertex.data = medians, vertex.cP = marker_palette(ncol(medians)), vertex.scale = TRUE, layout = layout, edge.lty = 1, mark.groups = codebook_list, mark.col = alpha(cluster_cols,0.4), mark.border = NA, mark.shape=1)
legend(x=1.2,y=1.1,legend=lineage_channels, col=marker_cols, pch=19, cex=0.7, bty=\&quot;n\&quot;, y.intersp=0.8)
legend(x=-1.8,y=1.1,legend=paste(\&quot;cluster\&quot;, 1:nclust), col=cluster_cols, pch=19, cex=0.7, bty=\&quot;n\&quot;, y.intersp=0.8)</code></pre>
<div class="figure">
<img src="https://i.imgur.com/J9hQvVk.png" alt="MST with SOM codebook clustering" />
<p class="caption">MST with SOM codebook clustering</p>
</div>
<p>Note that this visualization was developed for flow cytometry where fewer markers are measured. For higher-dimensionality data, such as from the CyTOF, it can be difficult to discern the expression of the individual markers, so selecting a subset representing major parent populations could be useful.</p>
</div>
</div>
</div>
<div id="analysis-tool-overview" class="section level1">
<h1><span class="header-section-number">4</span> Analysis tool overview</h1>
<table>
<colgroup>
<col width="9%" />
<col width="33%" />
<col width="56%" />
</colgroup>
<thead>
<tr class="header">
<th>Tool</th>
<th>Purpose</th>
<th>Reference</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>CytoRFD</td>
<td>Cell distance measurement</td>
<td><a href="https://doi.org/10.1109/EMBC.2016.7591260">Pouyan et al. (2016)</a></td>
</tr>
<tr class="even">
<td>ACDC</td>
<td>Cell type assignment</td>
<td><a href="https://doi.org/10.1093/bioinformatics/btx054">Lee et al. (2017)</a></td>
</tr>
<tr class="odd">
<td>CytoCompare</td>
<td>Cell type assignment</td>
<td><a href="https://doi.org/10.1016/j.ymeth.2017.09.005">Platon et al. (2018)</a></td>
</tr>
<tr class="even">
<td>DeepCyTOF</td>
<td>Cell type assignment</td>
<td><a href="https://doi.org/10.1093/bioinformatics/btx448">Li et al. (2017)</a></td>
</tr>
<tr class="odd">
<td>CyTOF-Linear-Classifier</td>
<td>Cell type assignment</td>
<td><a href="https://doi.org/10.1002/cyto.a.23738">Abdelaal et al. (2019)</a></td>
</tr>
<tr class="even">
<td>HiPPO and PANDA</td>
<td>Cell type assignment</td>
<td><a href="https://doi.org/10.7287/peerj.preprints.2188v1">Pirrò et al. (2016)</a></td>
</tr>
<tr class="odd">
<td>Mondrian</td>
<td>Cell type assignment</td>
<td><a href="https://arxiv.org/abs/1711.07673">Ji et al. (2017)</a></td>
</tr>
<tr class="even">
<td>MP (Mondrian)</td>
<td>Cell type assignment</td>
<td><a href="https://doi.org/10.1101/414904">Ji et al. (2018)</a></td>
</tr>
<tr class="odd">
<td>xCell</td>
<td>Cell type assignment</td>
<td><a href="https://doi.org/10.1101/114165">Aran et al. (2017)</a></td>
</tr>
<tr class="even">
<td>cytometree</td>
<td>Cell type assignment</td>
<td><a href="https://doi.org/10.1002/cyto.a.23601">Commenges et al. (2018)</a></td>
</tr>
<tr class="odd">
<td>SCAFFoLD</td>
<td>Cell type assignment, cellular trajectory mapping</td>
<td><a href="https://doi.org/10.1126/science.1259425">Spitzer et al. (2015)</a></td>
</tr>
<tr class="even">
<td>Statistical SCAFFoLD</td>
<td>Cell type assignment, cellular trajectory mapping, differential abundance analysis</td>
<td><a href="https://doi.org/10.1016/j.cell.2016.12.022">Spitzer et al. (2017)</a></td>
</tr>
<tr class="odd">
<td>DREMI/DREVI</td>
<td>Cellular trajectory detection</td>
<td><a href="https://doi.org/10.1126/science.1250689">Krishnaswamy et al. (2014)</a></td>
</tr>
<tr class="even">
<td>Wanderlust</td>
<td>Cellular trajectory detection</td>
<td><a href="https://doi.org/10.1016/j.cell.2014.04.005">Bendall et al. (2014)</a></td>
</tr>
<tr class="odd">
<td>ECLAIR</td>
<td>Cellular trajectory detection</td>
<td><a href="https://doi.org/10.1093/nar/gkw452">Giecold et al. (2016)</a></td>
</tr>
<tr class="even">
<td>TreeTop</td>
<td>Cellular trajectory detection</td>
<td><a href="https://doi.org/10.1101/200923">Macnair et al. (2017)</a></td>
</tr>
<tr class="odd">
<td>Wishbone</td>
<td>Cellular trajectory detection</td>
<td><a href="https://doi.org/10.1038/nbt.3569">Setty et al. (2016)</a></td>
</tr>
<tr class="even">
<td>PP-SLIDE</td>
<td>Cellular trajectory detection (prediction of precursor cells)</td>
<td><a href="https://doi.org/10.1016/j.celrep.2017.06.087">Cavrois et al. (2017)</a></td>
</tr>
<tr class="odd">
<td>ACCENSE</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1073/pnas.1321405111">Shekhar et al. (2014)</a></td>
</tr>
<tr class="even">
<td>DensVM</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1038/ni.3006">Becher et al. (2014)</a></td>
</tr>
<tr class="odd">
<td>FlowSOM</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1002/cyto.a.22625">Van Gassen et al. (2015)</a></td>
</tr>
<tr class="even">
<td>PhenoGraph</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1016/j.cell.2015.05.047">Levine et al. (2015)</a></td>
</tr>
<tr class="odd">
<td>SPADE</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1038/nbt.1991">Qui et al. (2011)</a></td>
</tr>
<tr class="even">
<td>X-Shift</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1038/nmeth.3863">Samusik et al. (2016)</a></td>
</tr>
<tr class="odd">
<td>ClusterX</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1371/journal.pcbi.1005112">Chen et al. (2016)</a></td>
</tr>
<tr class="even">
<td>FlowGrid</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1101/394189">Ye &amp; Ho (2018)</a></td>
</tr>
<tr class="odd">
<td>Rclusterpp</td>
<td>Clustering</td>
<td><a href="https://github.com/nolanlab/Rclusterpp">Nolan Lab GitHub</a></td>
</tr>
<tr class="even">
<td>CCAST</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1371/journal.pcbi.1003664">Anhang et al. (2014)</a></td>
</tr>
<tr class="odd">
<td>Convex clustering</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1111/rssb.12226">Radchenko &amp; Mukherjee (2017)</a></td>
</tr>
<tr class="even">
<td>CytoSPADE</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1093/bioinformatics/bts425">Linderman et al. (2012)</a></td>
</tr>
<tr class="odd">
<td>densityCUT</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1093/bioinformatics/btw227">Ding et al. (2016)</a></td>
</tr>
<tr class="even">
<td>DEPECHE</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1101/396135">Theorell et al. (2018)</a></td>
</tr>
<tr class="odd">
<td>flowMEANS</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1002/cyto.a.21007">Aghaeepour et al. (2011)</a></td>
</tr>
<tr class="even">
<td>immunoCLUST</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1002/cyto.a.22626">Sörensen et al. (2015)</a></td>
</tr>
<tr class="odd">
<td>OpenCyto</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1371/journal.pcbi.1003806">Finak et al. (2014)</a></td>
</tr>
<tr class="even">
<td>PAC-MAN</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1371/journal.pcbi.1005875">Pezzotti et al. (2017)</a></td>
</tr>
<tr class="odd">
<td>SOPHE</td>
<td>Clustering</td>
<td><a href="https://doi.org/10.1002/cyto.a.22704">Zaunders et al. (2015)</a></td>
</tr>
<tr class="even">
<td>Citrus</td>
<td>Clustering, differential abundance analysis</td>
<td><a href="https://doi.org/10.1073/pnas.1408792111">Bruggner et al. (2014)</a></td>
</tr>
<tr class="odd">
<td>Cydar</td>
<td>Clustering, differential abundance analysis</td>
<td><a href="https://doi.org/10.1038/nmeth.4295">Lun et al. (2017)</a></td>
</tr>
<tr class="even">
<td>diffcyt</td>
<td>Clustering, differential abundance analysis</td>
<td><a href="https://doi.org/10.1101/349738">Weber et al. (2018)</a></td>
</tr>
<tr class="odd">
<td>RchyOptimyx</td>
<td>Clustering, simulating manual gating</td>
<td><a href="https://doi.org/10.1002/cyto.a.22209">Aghaeepour et al. (2012)</a> and <a href="https://doi.org/10.1093/bioinformatics/btt770">O’Neill et al. (2014)</a></td>
</tr>
<tr class="even">
<td>MAGAN</td>
<td>Comparison of different types of measurements</td>
<td><a href="https://arxiv.org/abs/1803.00385">Amodio &amp; Krishnaswamy (2018)</a></td>
</tr>
<tr class="odd">
<td>MetaCYTO</td>
<td>Comparison of different types of measurements</td>
<td><a href="https://doi.org/10.1016/j.celrep.2018.07.003">Hu et al. (2018)</a></td>
</tr>
<tr class="even">
<td>Sensitivity Analysis</td>
<td>Covariance of markers</td>
<td><a href="https://arxiv.org/abs/1308.2329">G’Sell et al. (2013)</a></td>
</tr>
<tr class="odd">
<td>Stochastic Bayes</td>
<td>Covariance of markers</td>
<td><a href="https://arxiv.org/abs/1509.01631">Knowles (2015)</a></td>
</tr>
<tr class="even">
<td>CellCnn</td>
<td>Detection of disease-associated cell subsets</td>
<td><a href="https://doi.org/10.1038/ncomms14825">Arvanti &amp; Claassen (2017)</a></td>
</tr>
<tr class="odd">
<td>t-SNE</td>
<td>Dimensionality reduction</td>
<td><a href="https://www.jmlr.org/papers/v9/vandermaaten08a.html">van der Maaten et al. (2008)</a></td>
</tr>
<tr class="even">
<td>PCA</td>
<td>Dimensionality reduction</td>
<td></td>
</tr>
<tr class="odd">
<td>A-tSNE</td>
<td>Dimensionality reduction</td>
<td><a href="https://doi.org/10.1109/TVCG.2016.2570755">Li et al. (2017)</a></td>
</tr>
<tr class="even">
<td>CytoBinning</td>
<td>Dimensionality reduction</td>
<td><a href="https://doi.org/10.1101/321893">Shen et al. (2018)</a></td>
</tr>
<tr class="odd">
<td>destiny</td>
<td>Dimensionality reduction</td>
<td><a href="https://doi.org/10.1093/bioinformatics/btv715">Angerer et al. (2015)</a></td>
</tr>
<tr class="even">
<td>Cluster‐to‐Gate</td>
<td>Gating hierarchy generation</td>
<td><a href="https://doi.org/10.1002/cyto.a.23577">Yang &amp; Qui (2018)</a></td>
</tr>
<tr class="odd">
<td>Hypergate</td>
<td>Gating hierarchy generation</td>
<td><a href="https://doi.org/10.1093/bioinformatics/bty491">Becht et al. (2018)</a></td>
</tr>
<tr class="even">
<td>CytoBackBone</td>
<td>Improving marker selection</td>
<td><a href="https://github.com/tchitchek-lab/CytoBackBone">Tchitchek Lab GitHub</a></td>
</tr>
<tr class="odd">
<td>QFMatch</td>
<td>Matching of cell clusters between samples</td>
<td><a href="https://doi.org/10.1038/s41598-018-21444-4">Orlova et al. (2018)</a></td>
</tr>
<tr class="even">
<td>Hierarchical Bayesian</td>
<td>Network modeling</td>
<td><a href="https://doi.org/10.4137/CIN.S13984">Mitra et al. (2014)</a></td>
</tr>
<tr class="odd">
<td>SCENERY</td>
<td>Network modeling</td>
<td><a href="https://doi.org/10.1093/nar/gkx448">Papoutsoglou et al. (2017)</a></td>
</tr>
<tr class="even">
<td>CellCycleTRACER</td>
<td>Normalization of cell-cycle and cell-volume effects</td>
<td><a href="https://doi.org/10.1038/s41467-018-03005-5">Rapsomaniki et al. (2018)</a></td>
</tr>
<tr class="odd">
<td>Batch-Free Encoding</td>
<td>Removal of batch effects</td>
<td><a href="https://doi.org/10.1101/380816">Shaham (2018)</a></td>
</tr>
<tr class="even">
<td>SAUCIE</td>
<td>Removal of batch effects, clustering, visualization</td>
<td><a href="https://doi.org/10.1101/237065">Amodio et al. (2018)</a></td>
</tr>
<tr class="odd">
<td>GateFinder</td>
<td>Simulating manual gating</td>
<td><a href="https://doi.org/10.1093/bioinformatics/bty430">Aghaeepour et al. (2018)</a></td>
</tr>
<tr class="even">
<td>PHATE</td>
<td>Visualization</td>
<td><a href="https://doi.org/10.1101/120378">Moon et al. (2018)</a></td>
</tr>
<tr class="odd">
<td>Radviz</td>
<td>Visualization</td>
<td><a href="https://doi.org/10.1002/cyto.b.21499">Abraham et al. (2016)</a></td>
</tr>
<tr class="even">
<td>Sconify</td>
<td>Visualization</td>
<td><a href="https://doi.org/10.1101/337485">Burns et al. (2018)</a></td>
</tr>
<tr class="odd">
<td>Cytosplore</td>
<td>Visualization</td>
<td><a href="https://doi.org/10.1111/cgf.12893">Höllt et al. (2016)</a></td>
</tr>
<tr class="even">
<td>CyteGuide</td>
<td>Visualization</td>
<td><a href="https://doi.org/10.1109/TVCG.2017.2744318">Höllt et al. (2017)</a></td>
</tr>
<tr class="odd">
<td>NetworkPainter</td>
<td>Visualization</td>
<td><a href="https://doi.org/10.1186/s12859-015-0602-4">Karr et al. (2015)</a></td>
</tr>
<tr class="even">
<td>SLIDE</td>
<td>Visualization</td>
<td><a href="https://doi.org/10.1016/j.ymeth.2015.07.008">Sen et al. (2015)</a></td>
</tr>
</tbody>
</table>
<div id="r-session-info-sessioninfo" class="section level4">
<h4><span class="header-section-number">4.0.0.1</span> R session info {sessioninfo}</h4>
<pre class="r"><code>R version 3.4.1 Patched (2017-08-31 r73172)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: OS X El Capitan 10.11.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] cydar_1.0.0                SummarizedExperiment_1.6.3 DelayedArray_0.2.7         matrixStats_0.52.2         Biobase_2.36.2            
 [6] GenomicRanges_1.28.4       GenomeInfoDb_1.12.2        IRanges_2.10.2             S4Vectors_0.14.3           BiocGenerics_0.22.0       
[11] BiocParallel_1.10.1        flowCore_1.42.2            BiocInstaller_1.26.0      

loaded via a namespace (and not attached):
 [1] pcaPP_1.9-72            Rcpp_0.12.12            DEoptimR_1.0-8          compiler_3.4.1          plyr_1.8.4             
 [6] XVector_0.16.0          bitops_1.0-6            viridis_0.4.0           tools_3.4.1             zlibbioc_1.22.0        
[11] digest_0.6.12           viridisLite_0.2.0       tibble_1.3.4            gtable_0.2.0            lattice_0.20-35        
[16] rlang_0.1.2             Matrix_1.2-11           graph_1.54.0            shiny_1.0.5             mvtnorm_1.0-6          
[21] gridExtra_2.2.1         GenomeInfoDbData_0.99.0 cluster_2.0.6           grid_3.4.1              robustbase_0.92-7      
[26] R6_2.2.2                rrcov_1.4-3             plotrix_3.6-6           ggplot2_2.2.1           corpcor_1.6.9          
[31] htmltools_0.3.6         scales_0.5.0            xtable_1.8-2            mime_0.5                colorspace_1.3-2       
[36] httpuv_1.3.5            lazyeval_0.2.0          RCurl_1.95-4.8          munsell_0.4.3</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
